{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "# path hack for relative import in jupyter notebook\n",
    "import os\n",
    "import sys\n",
    "import time\n",
    "\n",
    "# LIBRARY GLOBAL MODS\n",
    "CELLTYPES = os.path.dirname(os.path.abspath(''))\n",
    "sys.path.append(CELLTYPES)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.datasets import load_digits\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.decomposition import PCA\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pickle\n",
    "import pandas as pd\n",
    "import umap\n",
    "import time\n",
    "%matplotlib inline\n",
    "\n",
    "sns.set(style='white', context='notebook', rc={'figure.figsize':(14,10)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib.colors import ListedColormap, BoundaryNorm, NoNorm, Normalize\n",
    "from matplotlib.colorbar import ColorbarBase\n",
    "from matplotlib.cm import get_cmap, ScalarMappable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# need to set cmaps to be used before importing multicell_replot \n",
    "#   as it uses proplot which modifies matplotlib\n",
    "\n",
    "CMAP_SPECTRALR = get_cmap('Spectral_r')\n",
    "GLOBAL_DPI = 450"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from utils.file_io import RUNS_FOLDER"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "NOTEBOOK_OUTDIR = RUNS_FOLDER + os.sep + 'explore' + os.sep + 'nb_alignUMAP'\n",
    "os.makedirs(NOTEBOOK_OUTDIR, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Setting to skip UMAP embedding and plots (slow part of notebook)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "NOTEBOOK_SPEEDRUN = False"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TODO: Import reloading while retaining variables in memory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import importlib\n",
    "#importlib.reload(multicell.multicell_replot)\n",
    "\n",
    "#import inspect\n",
    "#print(inspect.getsource(multicell.multicell_replot.replot_scatter_dots))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions: Non-Plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def reshape_totalspins(tallstate, num_cells, num_genes):\n",
    "    widestate = tallstate.reshape((num_genes, num_cells), order='F')\n",
    "    return widestate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_X_multi(manyruns_paths, subsample_ndim=None, subsample_ensemble=None, smod_last=True, use_01=True):\n",
    "    # any asserts and simple variable settings\n",
    "    if smod_last:\n",
    "        smod = '_last' # oldstyle\n",
    "    else:\n",
    "        smod = ''      # newstyle\n",
    "    \n",
    "    # get data dimensionality and set system size variables\n",
    "    ref_fpath_state = manyruns_paths[0] + os.sep + 'aggregate' + os.sep + 'X_aggregate%s.npz' % smod\n",
    "    ref_X = np.load(ref_fpath_state)['arr_0'].T  # umap wants transpose\n",
    "    npts, ndim = ref_X.shape\n",
    "    print(\"npts, ndim\", npts, ndim)\n",
    "    \n",
    "    if subsample_ensemble is not None:\n",
    "        assert 0 < subsample_ensemble <= npts\n",
    "        nn = subsample_ensemble\n",
    "    else: \n",
    "        nn = npts\n",
    "        \n",
    "    if subsample_ndim is not None:\n",
    "        assert 0 < subsample_ndim <= ndim\n",
    "        kk = subsample_ndim\n",
    "    else:\n",
    "        kk = ndim\n",
    "    \n",
    "    # Step 1: generate X_multi\n",
    "    X_multi = np.zeros((len(manyruns_paths), nn, kk), dtype=int)\n",
    "    for j, manyruns_path in enumerate(manyruns_paths):\n",
    "\n",
    "        #gamma_val = gamma_list[j]\n",
    "\n",
    "        agg_dir = manyruns_path + os.sep + 'aggregate'\n",
    "        fpath_state = agg_dir + os.sep + 'X_aggregate%s.npz' % smod\n",
    "        fpath_energy = agg_dir + os.sep + 'X_energy%s.npz' % smod\n",
    "        fpath_pickle = manyruns_path + os.sep + 'multicell_template.pkl'\n",
    "\n",
    "        X = np.load(fpath_state)['arr_0'].T  # umap wants transpose\n",
    "        X_energies = np.load(fpath_energy)['arr_0'].T  # umap wants transpose (?)\n",
    "        with open(fpath_pickle, 'rb') as pickle_file:\n",
    "            multicell_template = pickle.load(pickle_file)  # unpickling multicell object\n",
    "\n",
    "        #print('accessing', j, manyruns_path)\n",
    "        X_multi[j, :, :] = X[0:nn, 0:kk]\n",
    "    \n",
    "    if use_01:\n",
    "        X_multi = (1 + X_multi) / 2.0\n",
    "        X_multi = X_multi.astype(int)\n",
    "        \n",
    "    num_cells = multicell_template.num_cells\n",
    "    num_genes = multicell_template.num_genes\n",
    "    return X_multi, num_cells, num_genes\n",
    "\n",
    "\n",
    "def gather_data_then_alignedUMAP(\n",
    "    manyruns_paths, manyruns_dirnames, umap_kwargs, \n",
    "    skip_alignment=False, subsample_ndim=None, subsample_ensemble=None, \n",
    "    smod_last=True, use_01=True):\n",
    "    \"\"\"\n",
    "    All of these arguments are prepared in a dictionary \n",
    "        settings_aligned_umap_fig5_v1\n",
    "        settings_aligned_umap_fig5_v2\n",
    "        etc.\n",
    "    and chosen later in the notebook settings_alignment_chosen\n",
    "    \"\"\"\n",
    "    # TODO once this fn done, need to propogate data_subdict object through nb\n",
    "    #  - it should replace aligned_mapper\n",
    "    #  - should also integrate it with alignment settings dict somehow\n",
    "    # TODO should have class or validation function for data_subdict init\n",
    "    # TODO integrate side data like num unqieu states (analagous to energy)\n",
    "    \n",
    "    if smod_last:\n",
    "        smod = '_last'\n",
    "    else:\n",
    "        smod = ''\n",
    "\n",
    "    def build_data_subdict_from_alignedUMAP(data_subdict, X, embedding):\n",
    "        \"\"\"\n",
    "        data_subdict: partially complete data_subdict (path and label keys)\n",
    "        X:            raw data 2d arr (num_runs x num_dimensions)\n",
    "        embedding:    embedded data (num_runs x num_lowdim)\n",
    "        \"\"\"\n",
    "        # Reload data_subdict source files\n",
    "        #  note: we transpose raw data, energies for umap/other functions\n",
    "        manyruns_path = data_subdict['path']\n",
    "        agg_dir = manyruns_path + os.sep + 'aggregate'\n",
    "        fpath_state = agg_dir + os.sep + 'X_aggregate%s.npz' % smod\n",
    "        fpath_energy = agg_dir + os.sep + 'X_energy%s.npz' % smod\n",
    "        fpath_pickle = manyruns_path + os.sep + 'multicell_template.pkl'\n",
    "        X = np.load(fpath_state)['arr_0'].T          \n",
    "        X_energies = np.load(fpath_energy)['arr_0'].T\n",
    "        with open(fpath_pickle, 'rb') as pickle_file:\n",
    "            # unpickling multicell object\n",
    "            multicell_template = pickle.load(pickle_file)  \n",
    "        \n",
    "        # Fill out the rest of data_subdict keys        \n",
    "        num_runs, total_spins = X.shape\n",
    "        data_subdict['data'] = X\n",
    "        data_subdict['index'] = list(range(num_runs))\n",
    "        data_subdict['energies'] = np.round(X_energies, decimals=6)\n",
    "        data_subdict['num_runs'] = num_runs\n",
    "        data_subdict['total_spins'] = total_spins\n",
    "        data_subdict['multicell_template'] = multicell_template\n",
    "        data_subdict['algos'] = {\n",
    "            'umap': \n",
    "                {'reducer': None,\n",
    "                 'embedding': embedding}\n",
    "        }\n",
    "\n",
    "        # Add extra analysis data used within the notebook\n",
    "        X_nunique = get_unique_states_over_ensemble(\n",
    "            X, \n",
    "            multicell_template.num_cells, \n",
    "            multicell_template.num_genes\n",
    "        )\n",
    "        data_subdict['nunique'] = X_nunique.T  # note transpose\n",
    "        assert (\n",
    "            data_subdict['energies'].shape[0] == \n",
    "            data_subdict['nunique'].shape[0])\n",
    "        \n",
    "        return data_subdict\n",
    "    \n",
    "    # Step 0: any asserts and simple variable settings\n",
    "    assert 'random_state' in umap_kwargs.keys()\n",
    "    \n",
    "    # Step 1: construct X_multi (tensor of tissue state ensembles)\n",
    "    X_multi, _, _ = build_X_multi(\n",
    "        manyruns_paths, \n",
    "        subsample_ndim=subsample_ndim, \n",
    "        subsample_ensemble=subsample_ensemble, \n",
    "        smod_last=smod_last, \n",
    "        use_01=use_01\n",
    "    )\n",
    "    nn = X_multi.shape[1]\n",
    "    \n",
    "    # Step 2: perform alignedUMAP\n",
    "    aligned_mapper = umap.AlignedUMAP(**umap_kwargs)\n",
    "    \n",
    "    if NOTEBOOK_SPEEDRUN:\n",
    "        aligned_mapper.embeddings_ = [None for n in range(len(manyruns_paths))]\n",
    "    elif skip_alignment:\n",
    "        print('Warning - skipping alignment step...')\n",
    "        mappers = [\n",
    "            umap.UMAP(**umap_kwargs).fit(X_multi[n, :, :])\n",
    "            for n in range(len(manyruns_paths))\n",
    "        ]\n",
    "        embeddings = [mappers[n].embedding_ for n in range(len(manyruns_paths))]\n",
    "        aligned_mapper.embeddings_ = embeddings\n",
    "    else:\n",
    "        # UMAP aligned needs a 'relation dict' for the 'time varying' dataset\n",
    "        #   our relation is that each traj maps to itself (in time) -- constant relation\n",
    "        constant_dict = {i: i for i in range(nn)}\n",
    "        constant_relations = [constant_dict for i in range(len(manyruns_paths)-1)]\n",
    "        print('Beginning AlignedUMAP...')\n",
    "        aligned_mapper.fit(X_multi, relations=constant_relations)\n",
    "        print('...done')\n",
    "        \n",
    "    # Step 3: Convert to data_subdict format\n",
    "    # TODO - implement saving\n",
    "    embedded_datasets = {i: {'label': manyruns_dirnames[i],\n",
    "                             'path': manyruns_paths[i]}\n",
    "                         for i in range(len(manyruns_dirnames))}\n",
    "    \n",
    "    for idx in range(len(manyruns_dirnames)):\n",
    "        embedded_datasets[idx] = build_data_subdict_from_alignedUMAP(\n",
    "            embedded_datasets[idx], \n",
    "            X_multi[idx,:,:], \n",
    "            aligned_mapper.embeddings_[idx]\n",
    "        )\n",
    "        #fpath = manyruns_paths[idx] + os.sep + 'dimreduce' \\\n",
    "        #        + os.sep + 'dimreduce%s.z' % fmod\n",
    "        #save_dimreduce_object(datasets[idx], fpath)  # save to file (joblib)\n",
    "    \n",
    "    return embedded_datasets\n",
    "\n",
    "\n",
    "def tissue_to_num_unique_cells(X, num_cells, num_genes):\n",
    "    if X.shape != (num_genes, num_cells):\n",
    "        # perform the reshape\n",
    "        assert len(X) == num_genes * num_cells\n",
    "        X = reshape_totalspins(X, num_cells, num_genes)\n",
    "        \n",
    "    tissue_unique_cols = np.unique(X, axis=1)\n",
    "    num_unique = tissue_unique_cols.shape[1]\n",
    "    return num_unique, tissue_unique_cols\n",
    "\n",
    "\n",
    "def get_unique_states_over_ensemble(ensemble_array, num_cells, num_genes):\n",
    "    # TODO track with X_energies somehow...\n",
    "    num_ensemble = ensemble_array.shape[0]\n",
    "    carray = np.zeros(num_ensemble, dtype=int)\n",
    "    for idx in range(num_ensemble):\n",
    "        tissue_arr = ensemble_array[idx, :]\n",
    "        num_unique, unique_states = tissue_to_num_unique_cells(tissue_arr, num_cells, num_genes)\n",
    "        carray[idx] = num_unique\n",
    "    return carray"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_PCA_to_embedded_datasets(embedded_datasets):\n",
    "    for key in embedded_datasets.keys():\n",
    "        data_subdict = embedded_datasets[key]\n",
    "        assert 'pca' not in data_subdict['algos'].keys()\n",
    "        data_subdict['algos']['pca'] = {}\n",
    "        # get full PCA embedding\n",
    "        pca_full = PCA()\n",
    "        embedding = pca_full.fit_transform(data_subdict['data'])\n",
    "        data_subdict['algos']['pca']['embedding'] = embedding[:,:2]\n",
    "        # also store cum. sum. pca variance\n",
    "        exp_var_cumul = np.cumsum(pca_full.explained_variance_ratio_)\n",
    "        embedding = data_subdict['algos']['pca']['cumsum_x'] = \\\n",
    "            np.arange(1, exp_var_cumul.shape[0] + 1)\n",
    "        embedding = data_subdict['algos']['pca']['cumsum_y'] = \\\n",
    "            exp_var_cumul\n",
    "    return embedded_datasets\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions: Plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def axis_bounds(embedding):\n",
    "    left, right = embedding.T[0].min(), embedding.T[0].max()\n",
    "    bottom, top = embedding.T[1].min(), embedding.T[1].max()\n",
    "    adj_h, adj_v = (right - left) * 0.1, (top - bottom) * 0.1\n",
    "    return [left - adj_h, right + adj_h, bottom - adj_v, top + adj_v]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_fpdata_as_grid(\n",
    "    embedded_datasets, settings_alignment, fpdata_type, fpath, \n",
    "    gamma_selection=None, nrow=2, ncol=4, ext='.jpg', logy=False): \n",
    "    \"\"\"\n",
    "    Fixed point (FP) data forms: \n",
    "    - 'energies': energy array (1 dimensional -- 1 dim per fp)\n",
    "    - 'nunique': number of unique single cell states (1 dimensional -- 1 dim per fp)\n",
    "    Args:\n",
    "    - embedded_datasets: dict of data_subdicts objects\n",
    "    - settings_alignment: dict of manyruns info from jupyter notebook globals\n",
    "    - fpdata_type: 'energies' or 'nunique' or ['energies', 'nunique']\n",
    "    - gamma_selection: list of integers to choose a subset of data from settings_alignment['manyruns_paths']\n",
    "    \"\"\"\n",
    "    \n",
    "    VALID_FPDATA =['energies', 'nunique']\n",
    "    AXIS_LABELS = {\n",
    "        'nunique': 'Num. unique cell states',\n",
    "        'energies': 'Energy',\n",
    "    }\n",
    "    \n",
    "    gamma_values = settings_alignment['gamma_list']\n",
    "    manyruns_paths = settings_alignment['manyruns_paths']\n",
    "    #nn = settings_alignment['alignment_wrapper_kwargs']['subsample_ensemble']\n",
    "    if gamma_selection is None:\n",
    "        gamma_selection = settings_alignment['subplot_selection']\n",
    "    assert len(gamma_selection) == nrow * ncol\n",
    "    \n",
    "    def map_str_to_data(fpdata_str, gamma_idx):\n",
    "        assert fpdata_str in VALID_FPDATA\n",
    "        if fpdata_str == 'energies':\n",
    "            fpdata = embedded_datasets[gamma_idx][fpdata_str][:,0]\n",
    "        else:\n",
    "            fpdata = embedded_datasets[gamma_idx][fpdata_str]\n",
    "        \n",
    "        axis_label = AXIS_LABELS[fpdata_str]  \n",
    "        return fpdata, axis_label\n",
    "    \n",
    "    # determine which plot mode (1,2) is to be used based on number of fpdata_types provided\n",
    "    assert isinstance(fpdata_type, (str, list))\n",
    "    if isinstance(fpdata_type, (list)):\n",
    "        if len(fpdata_type) == 1:\n",
    "            assert isinstance(fpdata_type, (str))\n",
    "            fpdata_str = fpdata_type[0]\n",
    "            mode = '1'            \n",
    "        else:\n",
    "            assert len(fpdata_type) == 2\n",
    "            plot_mode = '2'\n",
    "    else:\n",
    "        fpdata_str = fpdata_type\n",
    "        plot_mode = '1'\n",
    "\n",
    "    if plot_mode == '1':\n",
    "        assert fpdata_str in VALID_FPDATA\n",
    "    \n",
    "        # TODO smarter bin settings: set after load data, observe statistics?\n",
    "        # TODO bins_settings = np.arange(0.5,40,1)  \n",
    "        data_arrays = [0] * len(gamma_selection)\n",
    "        if fpdata_type == 'energies':\n",
    "            bins_settings = 40\n",
    "            color = 'coral'\n",
    "            xlabel = 'Energy'\n",
    "            for idx, idx_selected in enumerate(gamma_selection):\n",
    "                data_arrays[idx] = embedded_datasets[idx_selected][fpdata_type][:, 0]\n",
    "        else:\n",
    "            max_count = 0\n",
    "            for idx, idx_selected in enumerate(gamma_selection):\n",
    "                data_arrays[idx] = embedded_datasets[idx_selected][fpdata_type]\n",
    "                max_count = max(max_count, np.max(data_arrays[idx]))                \n",
    "            print('max. num. s.c. types:', max_count)\n",
    "            bin_max = max(40, max_count + 1)\n",
    "            bins_settings = np.arange(0.5,bin_max,1)\n",
    "            color = 'deepskyblue'\n",
    "            xlabel = 'Number of unique cell states'\n",
    "\n",
    "        # plot settings\n",
    "        if logy:\n",
    "            fpath += '_logy'\n",
    "\n",
    "        # prints\n",
    "        print('Picks for plot_fpdata_as_grid():')\n",
    "        print([(i, gamma_values[i]) for i in gamma_selection])\n",
    "\n",
    "        # Step 2: plot as grid of histograms\n",
    "        fig, axs = plt.subplots(nrow, ncol, figsize=(18*0.68, 12*0.68))\n",
    "        for idx, ax in enumerate(axs.flatten()):\n",
    "            gamma_index = gamma_selection[idx]\n",
    "            gamma_value = gamma_values[gamma_index]\n",
    "            ptitle = r'%d: idx=%d, $\\gamma=%.3f$' % (\n",
    "                idx, gamma_index, gamma_value)\n",
    "            ax.set_title(ptitle, fontsize=12)\n",
    "            ax.hist(data_arrays[idx], bins=bins_settings, color=color)\n",
    "            if logy:\n",
    "                ax.set_yscale('log')\n",
    "\n",
    "        # Step 3: global axis labels\n",
    "        offset = 0.005\n",
    "        fig.text(0.5, offset, xlabel, ha='center')\n",
    "        fig.text(offset, 0.5, 'Number of fixed points', \n",
    "                 va='center', rotation='vertical')\n",
    "\n",
    "        #ax.axis(ax_bound)\n",
    "        #ax.set(xticks=[], yticks=[])\n",
    "\n",
    "    else:\n",
    "        assert plot_mode == '2'\n",
    "        data_x_str = fpdata_type[0]\n",
    "        data_y_str = fpdata_type[1]\n",
    "        assert data_x_str in VALID_FPDATA\n",
    "        assert data_y_str in VALID_FPDATA\n",
    "        \n",
    "        fig, axs = plt.subplots(nrow, ncol, figsize=(18*0.68, 12*0.68))\n",
    "        for idx, ax in enumerate(axs.flatten()):\n",
    "            gamma_index = gamma_selection[idx]\n",
    "            gamma_value = gamma_values[gamma_index]\n",
    "        \n",
    "            data_x, label_x = map_str_to_data(data_x_str, gamma_index)\n",
    "            data_y, label_y = map_str_to_data(data_y_str, gamma_index)\n",
    "            ax.scatter(data_x, data_y)\n",
    "            ax.set_title('$\\gamma=%.2f$ (idx %d)' % (gamma_value, gamma_index),\n",
    "                         fontsize=12)\n",
    "            \n",
    "        # global axis labels\n",
    "        offset = 0.005\n",
    "        fig.text(0.5, offset, label_x, ha='center')\n",
    "        fig.text(offset, 0.5, label_y, va='center', rotation='vertical')\n",
    "        #plt.title(r'%s vs %s for $\\gamma=%.2f$ (idx %d)' % \n",
    "        #              (data_x_str, data_y_str, gamma_value, gamma_index))\n",
    "        \n",
    "    plt.tight_layout(pad=1.2, w_pad=0.5, h_pad=1.0)\n",
    "    if isinstance(ext, list):\n",
    "        for ext_str in ext:\n",
    "            assert ext_str[0] == '.'\n",
    "            plt.savefig(fpath + ext_str, dpi=300)\n",
    "    else: \n",
    "        plt.savefig(fpath + ext, dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def annotate_a_point(ax, embedding, agg_idx, ss=10, zorder=4, addtext=False):\n",
    "    x, y = embedding[agg_idx, :]\n",
    "    ax.scatter(x, y, s=ss, facecolors='none', edgecolors='k',\n",
    "              path_effects=[PathEffects.withStroke(linewidth=2, foreground=\"w\")],\n",
    "              zorder=zorder)\n",
    "    if addtext:\n",
    "        ax.annotate('%d' % agg_idx, (x, y), \n",
    "                    xytext=(3, 3),\n",
    "                    textcoords='offset points',\n",
    "                    path_effects=[PathEffects.withStroke(linewidth=2, foreground=\"w\")],\n",
    "                    zorder=zorder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def transform_embedding(embedding, X_flip, X_rot):\n",
    "    degrees = X_rot * (np.pi / 180)\n",
    "    transformed_embedding = np.copy(embedding)\n",
    "    x = transformed_embedding[:, 0]\n",
    "    y = transformed_embedding[:, 1]\n",
    "    if X_flip:\n",
    "        x = -1 * x\n",
    "    transformed_embedding[:, 0] = np.cos(degrees) * x - np.sin(degrees) * y\n",
    "    transformed_embedding[:, 1] = np.sin(degrees) * x + np.cos(degrees) * y\n",
    "    return transformed_embedding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.patheffects as PathEffects\n",
    "\n",
    "def plot_aligned_panels_as_grid(\n",
    "    embedded_datasets, settings_alignment, fpath, \n",
    "    algo='umap',\n",
    "    gamma_selection=None, ext='.jpg', \n",
    "    ss=6, annotate_indices=[], addtext=True, rasterized=True,\n",
    "    fontsize=12,\n",
    "    c='energies', skip_cbar=False,\n",
    "    alpha=1.0, edges=False, black_edges=True,\n",
    "    X_flip=False, X_rot=0):\n",
    "    \"\"\"\n",
    "    Args: \n",
    "    - embedded_datasets: dict of data_subdicts (stores metadata, embedding)\n",
    "    - settings_alignment: dict of settings for the notebook,\n",
    "    - fpath: output path for plots (minus extension)\n",
    "    - algo: (str) key for data_subdict['algos'][algo] (e.g. 'umap', 'pca')\n",
    "    - gamma_selection: can override same key in settings_alignment\n",
    "    - ss: change marker size for scatter\n",
    "    - annotate_indices: list of indices to annotate on the embedding\n",
    "    - rasterized: If True, rasterize the scatter points at given dpi\n",
    "    - edges: if True, plot edges\n",
    "        - markers and edges separate calls to scatter\n",
    "        - alpha affects marker alpha only\n",
    "        - edge color default: face color (try switch to black)\n",
    "    Settings to consider changing:\n",
    "    - fontsize for title\n",
    "    Good defaults:\n",
    "    - ss=6, edges=True (with linewidth 0.1)\n",
    "    Notes:\n",
    "    - nunique cmap is based on \n",
    "        sns.color_palette(\"pastel\") or\n",
    "        sns.color_palette()\n",
    "        see https://seaborn.pydata.org/generated/seaborn.color_palette.html\n",
    "    \n",
    "    \"\"\"\n",
    "    # this line maintains cmap colors (matplotlib and proplot fighting)\n",
    "    sns.set(style='white', context='notebook')\n",
    "    dpi = GLOBAL_DPI\n",
    "    \n",
    "    # parse info from dict: settings_alignment\n",
    "    gamma_list = settings_alignment['gamma_list']\n",
    "    manyruns_paths = settings_alignment['manyruns_paths']\n",
    "    nn = settings_alignment['alignment_wrapper_kwargs']['subsample_ensemble']\n",
    "    assert nn is None\n",
    "    if gamma_selection is None:\n",
    "        gamma_selection = settings_alignment['subplot_selection']\n",
    "    \n",
    "    # parse info from dict: embedded_datasets (contains data_subdicts)\n",
    "    # TODO handler for embedding dimension > 2?\n",
    "    assert algo in embedded_datasets[0]['algos'].keys()\n",
    "    embeddings = [\n",
    "        transform_embedding(\n",
    "            embedded_datasets[idx]['algos'][algo]['embedding'][:, 0:2],\n",
    "            X_flip,\n",
    "            X_rot\n",
    "        )\n",
    "        for idx in range(len(settings_alignment_chosen['gamma_list']))\n",
    "    ]\n",
    "\n",
    "    # Part A: Figure initialization\n",
    "    # see https://matplotlib.org/stable/tutorials/intermediate/gridspec.html\n",
    "    num_rows = 2\n",
    "    num_cols = 4\n",
    "    if skip_cbar:\n",
    "        #fig = plt.figure(figsize=(5,5), dpi=dpi)\n",
    "        fig, axs = plt.subplots(\n",
    "            num_rows, num_cols, figsize=(10.24, 5.16), dpi=dpi)    # def: 12.24, 8.16\n",
    "    else:\n",
    "        cbar_horizontal = False\n",
    "        if cbar_horizontal:\n",
    "            fig, axs = plt.subplots(\n",
    "                (num_rows+1), num_cols, figsize=(10.24, 5.16), dpi=dpi,  # def: 12.24, 8.16\n",
    "                gridspec_kw={'height_ratios': [1,1,0.1]}\n",
    "            )\n",
    "            orientation = 'horizontal'\n",
    "            gs = axs[-1,-1].get_gridspec()\n",
    "            cbar_ax = fig.add_subplot(gs[-1, :])\n",
    "            for ax in axs[-1, :]:\n",
    "                ax.remove()\n",
    "        else:\n",
    "            fig, axs = plt.subplots(\n",
    "                num_rows, (num_cols+1), figsize=(10.24, 5.16), dpi=dpi,  # def: 12.24, 8.16 \n",
    "                gridspec_kw={'width_ratios': [1,1,1,1,0.1]}\n",
    "            )\n",
    "            orientation = 'vertical'\n",
    "            gs = axs[0, -1].get_gridspec()\n",
    "            cbar_ax = fig.add_subplot(gs[:, -1])\n",
    "            for ax in axs[:, -1]:\n",
    "                ax.remove()\n",
    "        cbar_ax.set(xticks=[], yticks=[])\n",
    "        \n",
    "    # Part B: Plot coloring rules (static, nunique, or energy)\n",
    "    if c is None:\n",
    "        fixed_color = 'gainsboro'  # '#5076b7', 'slategray', 'gainsboro'\n",
    "        cmap = None\n",
    "    else:\n",
    "        assert c in ['energies', 'nunique']\n",
    "        if c == 'energies':\n",
    "            #cmap_str = 'Spectral_r'\n",
    "            #cmap = get_cmap(cmap_str)\n",
    "            cmap = CMAP_SPECTRALR\n",
    "            \n",
    "            if not skip_cbar:\n",
    "                artificial_norm = Normalize(vmin=-1, vmax=1)\n",
    "                mappable = ScalarMappable(artificial_norm, cmap=cmap)\n",
    "                cbar = plt.colorbar(\n",
    "                    mappable,\n",
    "                    cax=cbar_ax,\n",
    "                    orientation=orientation,\n",
    "                )\n",
    "                cbar.set_label('Energy', fontsize=fontsize)\n",
    "                cbar.ax.tick_params(size=0)\n",
    "                if cbar_horizontal:\n",
    "                    cbar.ax.set_xticklabels([], fontsize=fontsize)\n",
    "                else:\n",
    "                    cbar.ax.set_yticklabels([], fontsize=fontsize)\n",
    "                #cbar_ax.set_xticks([])\n",
    "                #cbar_ax.set(xticks=[], yticks=[])\n",
    "            \n",
    "        else:\n",
    "            # create bounds for custom discrete cmap\n",
    "            #nunique = data_subdict['nunique']\n",
    "            bounds_upper = 22 #  np.max(nunique), or static if proportional cbar\n",
    "            bounds = [1, 2, 3, 4, 8, 15, bounds_upper]\n",
    "            bounds_midpts = [0.5 * (bounds[i] + bounds[i+1]) for i in range(len(bounds) - 1)]\n",
    "            ncolors = len(bounds) - 1\n",
    "            # refining bounds labels\n",
    "            bounds_labels = [''] * ncolors\n",
    "            for i in range(len(bounds) - 2):\n",
    "                lower = bounds[i]\n",
    "                upper = bounds[i+1] - 1\n",
    "                if lower == upper:\n",
    "                    bounds_labels[i] = '%d' % lower\n",
    "                else:\n",
    "                    bounds_labels[i] = '%d to %d' % (lower, upper)\n",
    "            #bounds_labels = ['%d to %d' % (bounds[i], bounds[i+1] - 1) for i in range(len(bounds) - 1)]\n",
    "            bounds_labels[-1] = r'$\\geq%d$' % bounds[-2]\n",
    "\n",
    "            spectral_auto = False\n",
    "            if spectral_auto:\n",
    "                #cmap_primitive = get_cmap('Spectral_r')\n",
    "                cmap_primitive = CMAP_SPECTRALR\n",
    "                norm_manual = (np.array(bounds_midpts) - bounds_midpts[0]) / (bounds_midpts[-1] - bounds_midpts[0])\n",
    "                print(norm_manual)\n",
    "                color_palette = [cmap_primitive(a) for a in np.linspace(0, 1, ncolors)]\n",
    "            else:\n",
    "                sns_pastel_indices = [0,9,2,3,4,8]\n",
    "                \n",
    "                #color_palette = [sns.color_palette(\"pastel\")[i] \n",
    "                #                 for i in sns_pastel_indices] \n",
    "                color_palette = [sns.color_palette()[i] \n",
    "                                 for i in sns_pastel_indices] \n",
    "                assert len(color_palette) >= ncolors\n",
    "                color_palette = color_palette[0:ncolors]\n",
    "            cmap = ListedColormap(color_palette)\n",
    "            \n",
    "            # create scatter array of colors\n",
    "            norm_nunique = BoundaryNorm(bounds, cmap.N, clip=True)\n",
    "            #color = cmap(norm(nunique))\n",
    "            \n",
    "            if not skip_cbar:\n",
    "                # this is alternative to gridspec:\n",
    "                #fig.subplots_adjust(right=0.8)\n",
    "                #cbar_ax = fig.add_axes([0.85, 0.15, 0.05, 0.7])\n",
    "                cbar = ColorbarBase(\n",
    "                    cbar_ax,\n",
    "                    cmap=cmap,\n",
    "                    norm=norm_nunique,\n",
    "                    #boundaries=[0] + bounds + [13],\n",
    "                    ticks=bounds_midpts,\n",
    "                    spacing='proportional',  # 'proportional', 'uniform'\n",
    "                    orientation=orientation)\n",
    "                if cbar_horizontal:\n",
    "                    cbar.ax.set_xticklabels(bounds_labels, fontsize=fontsize)\n",
    "                else:\n",
    "                    cbar.ax.set_yticklabels(bounds_labels, fontsize=fontsize)\n",
    "                cbar.set_label('Number of unique single cell states', fontsize=fontsize)\n",
    "    \n",
    "    # Part C: Edges (on scatterplot points) or not? \n",
    "    sc_kwargs = { \n",
    "        'cmap': cmap,\n",
    "        's': ss,\n",
    "        'rasterized': rasterized,\n",
    "        'alpha': alpha}\n",
    "    sc_kwargs['edgecolors'] = 'None'\n",
    "    if edges:\n",
    "        #sc_kwargs['edgecolors'] = 'None'\n",
    "        sc_kwargs_edge = sc_kwargs.copy()\n",
    "        sc_kwargs_edge.pop('edgecolors')\n",
    "        sc_kwargs_edge['alpha'] = 1.0\n",
    "        sc_kwargs_edge['facecolor'] = 'None'\n",
    "        sc_kwargs_edge['linewidths'] = 0.4  # 0.1  # ss / 20.0\n",
    "        \n",
    "    # Part D: Fill in each grid box (default - 2x4 panels)\n",
    "    ax_bound = axis_bounds(np.vstack(embeddings))\n",
    "\n",
    "    #for j, ax in enumerate(axs.flatten()):\n",
    "        #if j < len(gamma_selection):\n",
    "    \n",
    "    for j in range(num_rows * num_cols):\n",
    "        row_idx = j // num_cols\n",
    "        col_idx = j % num_cols\n",
    "        ax = axs[row_idx, col_idx] \n",
    "        i = gamma_selection[j]\n",
    "\n",
    "        data_subdict = embedded_datasets[i]\n",
    "        if c == 'energies':\n",
    "            color_arr = data_subdict['energies'][:, 0]\n",
    "            cnorm = Normalize(\n",
    "                vmin=np.min(color_arr), vmax=np.max(color_arr))\n",
    "            colors = cmap(cnorm(color_arr))\n",
    "        elif c == 'nunique':\n",
    "            color_arr = data_subdict['nunique']\n",
    "            #cnorm = Normalize(\n",
    "            #    vmin=np.min(color_arr), vmax=np.max(color_arr))\n",
    "            colors = cmap(norm_nunique(color_arr))\n",
    "        else:\n",
    "            colors = fixed_color\n",
    "\n",
    "        ax.set_title(gamma_list[i], fontsize=16, y=-0.17)\n",
    "\n",
    "        sc = ax.scatter(*embeddings[i].T, \n",
    "                        **sc_kwargs, \n",
    "                        c=colors,\n",
    "                        zorder=3)\n",
    "        # Use two scatter calls to get edges with different alpha\n",
    "        if edges:\n",
    "            if black_edges:\n",
    "                edgecolors = 'black'\n",
    "            else:\n",
    "                edgecolors = colors\n",
    "            sc2 = ax.scatter(*embeddings[i].T, \n",
    "                             **sc_kwargs_edge, \n",
    "                             edgecolors=edgecolors,\n",
    "                             zorder=2)\n",
    "\n",
    "        # annotate certain elements with empty circles\n",
    "        if isinstance(annotate_indices, dict):\n",
    "            # dict of lists for each gamma index\n",
    "            annotate_indices_specific = annotate_indices[i]\n",
    "            for agg_idx in annotate_indices_specific:\n",
    "                annotate_a_point(ax, embeddings[i], agg_idx, zorder=4, addtext=addtext)\n",
    "        else:\n",
    "            # assume its the same list for each gamma value\n",
    "            for agg_idx in annotate_indices:\n",
    "                annotate_a_point(ax, embeddings[i], agg_idx, zorder=4, addtext=addtext)\n",
    "\n",
    "        #ax.axis(ax_bound)\n",
    "        ax.set(xticks=[], yticks=[])\n",
    "\n",
    "    # Part E: Plot Save\n",
    "    plt.tight_layout()\n",
    "    if isinstance(ext, list):\n",
    "        for ext_str in ext:\n",
    "            assert ext_str[0] == '.'\n",
    "            plt.savefig(fpath + ext_str, dpi=dpi)\n",
    "    else: \n",
    "        plt.savefig(fpath + ext, dpi=dpi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_single_umap_panel_gridspec(\n",
    "    data_subdict, fpath, \n",
    "    c=None, ext='.jpg', ss=10, annotate_indices=[], addtext=True,\n",
    "    rasterized=True, skip_cbar=False, \n",
    "    alpha=1.0, edges=False,\n",
    "    X_flip=False, X_rot=0,\n",
    "    cbar_horizontal=True,\n",
    "    figsize=None):\n",
    "    \n",
    "    # this line maintains cmap colors (matplotlib and proplot fighting)\n",
    "    sns.set(style='white', context='notebook')\n",
    "    dpi = GLOBAL_DPI\n",
    "    \n",
    "    num_runs = data_subdict['num_runs']\n",
    "    label = data_subdict['label']\n",
    "    embedding = data_subdict['algos']['umap']['embedding']\n",
    "    assert embedding.shape[1] == 2\n",
    "    embedding = transform_embedding(embedding, X_flip, X_rot)\n",
    "    \n",
    "    if skip_cbar:\n",
    "        if figsize is None:\n",
    "            figsize = (5,5)\n",
    "        fig = plt.figure(figsize=figsize, dpi=dpi)\n",
    "    else:\n",
    "        if cbar_horizontal:\n",
    "            if figsize is None:\n",
    "                figsize = (5.5, 5)\n",
    "            fig, ax = plt.subplots(\n",
    "                2, 1, figsize=figsize, dpi=dpi,\n",
    "                gridspec_kw={'height_ratios': [12, 1]}\n",
    "            )\n",
    "            orientation = 'horizontal'\n",
    "        else:\n",
    "            # original figsize: (5, 5.5)\n",
    "            if figsize is None:\n",
    "                figsize = (5, 5.5)\n",
    "            fig, ax = plt.subplots(\n",
    "                1, 2, figsize=figsize, dpi=dpi, \n",
    "                gridspec_kw={'width_ratios': [12, 1]}\n",
    "            )\n",
    "            orientation = 'vertical'\n",
    "\n",
    "    if c is None:\n",
    "        color = '#5076b7'  # '#5076b7', 'slategray', 'gainsboro'\n",
    "        cmap = None\n",
    "    else:\n",
    "        assert c in ['energies', 'nunique']\n",
    "        if c == 'energies':\n",
    "            color = data_subdict['energies'][:, 0]  # range(num_runs)\n",
    "            cmap = 'Spectral_r'\n",
    "        else:\n",
    "            nunique = data_subdict['nunique']\n",
    "            \n",
    "            bounds_upper = 22 #  np.max(nunique), or static if proportional cbar\n",
    "            bounds = [1, 2, 3, 4, 8, 15, bounds_upper]\n",
    "            bounds_midpts = [0.5 * (bounds[i] + bounds[i+1]) for i in range(len(bounds) - 1)]\n",
    "            ncolors = len(bounds) - 1\n",
    "            # refining bounds labels\n",
    "            bounds_labels = [''] * ncolors\n",
    "            for i in range(len(bounds) - 2):\n",
    "                lower = bounds[i]\n",
    "                upper = bounds[i+1] - 1\n",
    "                if lower == upper:\n",
    "                    bounds_labels[i] = '%d' % lower\n",
    "                else:\n",
    "                    bounds_labels[i] = '%d to %d' % (lower, upper)\n",
    "            #bounds_labels = ['%d to %d' % (bounds[i], bounds[i+1] - 1) for i in range(len(bounds) - 1)]\n",
    "            bounds_labels[-1] = r'$\\geq%d$' % bounds[-2]\n",
    "\n",
    "            spectral_auto = False\n",
    "            if spectral_auto:\n",
    "                #cmap_primitive = get_cmap('Spectral_r')\n",
    "                cmap_primitive = CMAP_SPECTRALR\n",
    "                norm_manual = (np.array(bounds_midpts) - bounds_midpts[0]) / (bounds_midpts[-1] - bounds_midpts[0])\n",
    "                print(norm_manual)\n",
    "                color_palette = [cmap_primitive(a) for a in np.linspace(0, 1, ncolors)]\n",
    "            else:\n",
    "                #colors_bank = [cmap_primitive(a) for a in \n",
    "                #               [0.0, 0.1, 0.15, 0.25, 0.35, 0.55, 0.65, 0.75, 0.85, 1.0]]\n",
    "                \n",
    "                # 12 max for sns \"Paired\", manually rearrange the color list first\n",
    "                #sns_paired_indices = [0,1,2,3,9,8]\n",
    "                #sns_paired_indices = [0,1,2,3,7,6]\n",
    "                #color_palette = [sns.color_palette(\"Paired\")[i] \n",
    "                #                 for i in sns_paired_indices] \n",
    "                \n",
    "                #sns_set2_indices = [0,1,2,3,4,7]\n",
    "                #color_palette = [sns.color_palette(\"Set2\")[i] \n",
    "                #                 for i in sns_set2_indices] \n",
    "                \n",
    "                #sns_pastel_indices = [0,1,2,3,4,6]\n",
    "                #sns_pastel_indices = [0,1,2,3,4,9]\n",
    "                #sns_pastel_indices = [0,7,2,3,4,9]\n",
    "                #sns_pastel_indices = [0,9,2,3,4,7]\n",
    "                sns_pastel_indices = [0,9,2,3,4,8]\n",
    "                #color_palette = [sns.color_palette(\"pastel\")[i] \n",
    "                #                 for i in sns_pastel_indices] \n",
    "                color_palette = [sns.color_palette()[i] \n",
    "                                 for i in sns_pastel_indices]\n",
    "                \n",
    "                assert len(color_palette) >= ncolors\n",
    "                color_palette = color_palette[0:ncolors]\n",
    "                \n",
    "                \n",
    "            cmap = ListedColormap(color_palette)\n",
    "            \n",
    "            # create scatter array of colors\n",
    "            norm = BoundaryNorm(bounds, cmap.N, clip=True)\n",
    "            color = cmap(norm(nunique))\n",
    "            \n",
    "            # use grid spec to make two axes for this block\n",
    "            if not skip_cbar:\n",
    "                cbar = ColorbarBase(\n",
    "                    ax[1], \n",
    "                    cmap=cmap,\n",
    "                    norm=norm,\n",
    "                    #boundaries=[0] + bounds + [13],\n",
    "                    ticks=bounds_midpts,\n",
    "                    spacing='proportional',  # 'proportional', 'uniform'\n",
    "                    orientation=orientation)\n",
    "                if cbar_horizontal:\n",
    "                    cbar.ax.set_xticklabels(bounds_labels)\n",
    "                else:\n",
    "                    cbar.ax.set_yticklabels(bounds_labels)\n",
    "                cbar.set_label('Number of unique single cell states')\n",
    "            \n",
    "    sc_kwargs = {\n",
    "        'c': color, \n",
    "        'cmap': cmap,\n",
    "        's': ss,\n",
    "        'rasterized': rasterized,\n",
    "        'alpha': alpha        \n",
    "    }\n",
    "    sc_kwargs['edgecolors'] = 'None'\n",
    "    if edges:\n",
    "        sc_kwargs_edge = sc_kwargs.copy()\n",
    "        sc_kwargs_edge['alpha'] = 1.0\n",
    "        sc_kwargs_edge['facecolor'] = 'None'\n",
    "        sc_kwargs_edge['linewidths'] = 0.25\n",
    "        sc_kwargs_edge['edgecolors'] = 'black'\n",
    "            \n",
    "    if skip_cbar:\n",
    "        ax_sc = plt.gca()\n",
    "    else: \n",
    "        ax_sc = ax[0]\n",
    "    sc = ax_sc.scatter(\n",
    "        embedding[:, 0], embedding[:, 1], **sc_kwargs, zorder=3)\n",
    "    if edges:\n",
    "        sc2 = ax_sc.scatter(\n",
    "            embedding[:, 0], embedding[:, 1], **sc_kwargs_edge, zorder=2)\n",
    "    #ax_sc.set_aspect('equal', 'datalim')\n",
    "    ax_sc.set(xticks=[], yticks=[])\n",
    "    \n",
    "    # annotate certain elements with empty circles\n",
    "    annotate_ss = 15\n",
    "    if isinstance(annotate_indices, dict):\n",
    "        # dict of lists for each gamma index\n",
    "        annotate_indices_specific = annotate_indices[i]\n",
    "        for agg_idx in annotate_indices_specific:\n",
    "            annotate_a_point(ax_sc, embedding, agg_idx, ss=annotate_ss,\n",
    "                             zorder=4, addtext=addtext)\n",
    "    else:\n",
    "        # assume its the same list for each gamma value\n",
    "        for agg_idx in annotate_indices:\n",
    "            annotate_a_point(ax_sc, embedding, agg_idx, ss=annotate_ss, \n",
    "                             zorder=4, addtext=addtext)\n",
    "    \n",
    "    if c == 'energies' and not skip_cbar:\n",
    "        cbar = plt.colorbar(sc, cax=ax[1], orientation=orientation)\n",
    "        cbar.ax.tick_params(size=0)\n",
    "        #plt.colorbar(boundaries=np.arange(11)-0.5).set_ticks(np.arange(10))\n",
    "\n",
    "    # plt.title('UMAP projection of the %s dataset' % label, fontsize=24)\n",
    "    \n",
    "    plt.tight_layout()\n",
    "    if isinstance(ext, list):\n",
    "        for ext_str in ext:\n",
    "            assert ext_str[0] == '.'\n",
    "            plt.savefig(fpath + ext_str, dpi=dpi)\n",
    "    else: \n",
    "        plt.savefig(fpath + ext, dpi=dpi)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Specify settings and gen/load alignedUMAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "RANDOM_STATE = 20  #0 #100 #40 #20 #0  Note RANDOM_STATE = 20 for July 2021 fig. generation\n",
    "\n",
    "UMAP_KWARGS_DEFAULT = {\n",
    "    'random_state': RANDOM_STATE,  \n",
    "    'transform_seed': RANDOM_STATE,\n",
    "    'n_components': 2,\n",
    "    'metric': 'euclidean',\n",
    "    'init': 'spectral',\n",
    "    'unique': False,\n",
    "    'n_neighbors': 15,  # Default: 15\n",
    "    'min_dist': 0.1,    # Default: 0.1\n",
    "    'spread': 1.0,      # Default: 1.0\n",
    "    'n_epochs': 200,    # Default: 200\n",
    "}\n",
    "\n",
    "UMAP_KWARGS_ALT = {\n",
    "    'random_state': RANDOM_STATE,  \n",
    "    'transform_seed': RANDOM_STATE,\n",
    "    'n_components': 2,\n",
    "    'metric': 'euclidean',\n",
    "    'init': 'spectral',\n",
    "    'unique': False,\n",
    "    'n_neighbors': 15,  # Default: 15\n",
    "    'min_dist': 0.1,    # Default: 0.1\n",
    "    'spread': 1.0,      # Default: 1.0\n",
    "    'n_epochs': 200     # Default: 200\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO - consider also including a subset parameter for relationsdict (subset to 900 was the preJUly Fig5)\n",
    "\n",
    "# Settings used to generate Fig 4, 5, associated SI Figures\n",
    "gamma_list_v1 = [0.0, 0.05, 0.06, 0.07, 0.08, \n",
    "                 0.09, 0.10, 0.15, 0.20, 0.4, \n",
    "                 0.6, 0.8, 0.9, 1.0, 20.0]\n",
    "manyruns_dirnames_v1 = [\n",
    "    'Wrandom0_gamma%.2f_10k_periodic_fixedorderV3_p3_M100' % a \n",
    "    for a in gamma_list_v1]\n",
    "manyruns_paths_v1 = [RUNS_FOLDER + os.sep + 'multicell_manyruns' + os.sep + dirname for dirname in manyruns_dirnames_v1]\n",
    "settings_aligned_umap_fig5_v1 = {\n",
    "    'gamma_list': gamma_list_v1,\n",
    "    'manyruns_paths': manyruns_paths_v1,\n",
    "    'manyruns_dirnames': manyruns_dirnames_v1,\n",
    "    'umap_kwargs': UMAP_KWARGS_DEFAULT,\n",
    "    'alignment_wrapper_kwargs': \n",
    "        {'subsample_ndim': None, \n",
    "         'subsample_ensemble': None,  # 4000 originally\n",
    "         'smod_last': True,\n",
    "         'use_01': False,\n",
    "         'skip_alignment': False},\n",
    "    'subplot_selection': [0, 3, 4, 5, 7, 8, 9, 13]\n",
    "}\n",
    "\n",
    "gamma_list_v2 = [0.0, 0.07, 0.08, 0.09, 0.15, 0.20, 0.4, 1.0]\n",
    "manyruns_dirnames_v2 = [\n",
    "    'Wrandom0_gamma%.2f_10k_periodic_fixedorderV3_p3_M100' % a \n",
    "    for a in gamma_list_v2]\n",
    "manyruns_paths_v2 = [RUNS_FOLDER + os.sep + 'multicell_manyruns' + os.sep + dirname for dirname in manyruns_dirnames_v2]\n",
    "settings_aligned_umap_fig5_v2 = {\n",
    "    'gamma_list': gamma_list_v2,\n",
    "    'manyruns_paths': manyruns_paths_v2,\n",
    "    'manyruns_dirnames': manyruns_dirnames_v2,\n",
    "    'umap_kwargs': UMAP_KWARGS_ALT,\n",
    "    'alignment_wrapper_kwargs': \n",
    "        {'subsample_ndim': None, \n",
    "         'subsample_ensemble': None,\n",
    "         'smod_last': True,\n",
    "         'use_01': False,\n",
    "         'skip_alignment': False},\n",
    "    'subplot_selection': np.arange(8)    \n",
    "}\n",
    "\n",
    "gamma_list_v3 = gamma_list_v1\n",
    "manyruns_dirnames_v3 = [\n",
    "    'Wrandom0_gamma%.2f_10k_periodic_R1_p3_M100_machineEps' % a \n",
    "    for a in gamma_list_v3]\n",
    "manyruns_paths_v3 = [\n",
    "    RUNS_FOLDER + os.sep + 'multicell_manyruns' + os.sep + dirname \n",
    "    for dirname in manyruns_dirnames_v3]\n",
    "settings_aligned_umap_fig5_v3 = {\n",
    "    'gamma_list': gamma_list_v3,\n",
    "    'manyruns_paths': manyruns_paths_v3,\n",
    "    'manyruns_dirnames': manyruns_dirnames_v3,\n",
    "    'umap_kwargs': UMAP_KWARGS_DEFAULT,\n",
    "    'alignment_wrapper_kwargs': \n",
    "        {'subsample_ndim': None, \n",
    "         'subsample_ensemble': None,\n",
    "         'smod_last': True,\n",
    "         'use_01': False,\n",
    "         'skip_alignment': False},\n",
    "    'subplot_selection': [0, 3, 4, 5, 7, 8, 9, 13]    \n",
    "}\n",
    "\n",
    "\n",
    "settings_aligned_umap_W1 = {\n",
    "    'gamma_list': gamma_list_v1,\n",
    "    'umap_kwargs': UMAP_KWARGS_DEFAULT,\n",
    "    'alignment_wrapper_kwargs': \n",
    "        {'subsample_ndim': None, \n",
    "         'subsample_ensemble': None,\n",
    "         'smod_last': True,\n",
    "         'use_01': False,\n",
    "         'skip_alignment': False},\n",
    "    'subplot_selection': [0, 3, 4, 5, 7, 8, 9, 13]    \n",
    "}\n",
    "settings_aligned_umap_W1['manyruns_dirnames'] = [\n",
    "    'W1_gamma%.2f_10k_periodic_R1_p3_M100_machineEps' % a \n",
    "    for a in gamma_list_v1]\n",
    "settings_aligned_umap_W1['manyruns_paths'] = [\n",
    "    RUNS_FOLDER + os.sep + 'multicell_manyruns' + os.sep + dirname \n",
    "    for dirname in settings_aligned_umap_W1['manyruns_dirnames']]\n",
    "\n",
    "\n",
    "settings_aligned_umap_W2 = {\n",
    "    'gamma_list': gamma_list_v1,\n",
    "    'umap_kwargs': UMAP_KWARGS_DEFAULT,\n",
    "    'alignment_wrapper_kwargs': \n",
    "        {'subsample_ndim': None, \n",
    "         'subsample_ensemble': None,\n",
    "         'smod_last': True,\n",
    "         'use_01': False,\n",
    "         'skip_alignment': False},\n",
    "    'subplot_selection': [0, 3, 4, 5, 7, 8, 9, 13]    \n",
    "}\n",
    "settings_aligned_umap_W2['manyruns_dirnames'] = [\n",
    "    'W2_gamma%.2f_10k_periodic_R1_p3_M100_machineEps' % a \n",
    "    for a in gamma_list_v1]\n",
    "settings_aligned_umap_W2['manyruns_paths'] = [\n",
    "    RUNS_FOLDER + os.sep + 'multicell_manyruns' + os.sep + dirname \n",
    "    for dirname in settings_aligned_umap_W2['manyruns_dirnames']]\n",
    "\n",
    "\n",
    "settings_aligned_umap_W7 = {\n",
    "    'gamma_list': gamma_list_v1,\n",
    "    'umap_kwargs': UMAP_KWARGS_DEFAULT,\n",
    "    'alignment_wrapper_kwargs': \n",
    "        {'subsample_ndim': None, \n",
    "         'subsample_ensemble': None,\n",
    "         'smod_last': True,\n",
    "         'use_01': False,\n",
    "         'skip_alignment': False},\n",
    "    'subplot_selection': [0, 3, 4, 5, 7, 8, 9, 13]    \n",
    "}\n",
    "settings_aligned_umap_W7['manyruns_dirnames'] = [\n",
    "    'W7_gamma%.2f_10k_periodic_R1_p3_M100_machineEps' % a \n",
    "    for a in gamma_list_v1]\n",
    "settings_aligned_umap_W7['manyruns_paths'] = [\n",
    "    RUNS_FOLDER + os.sep + 'multicell_manyruns' + os.sep + dirname \n",
    "    for dirname in settings_aligned_umap_W7['manyruns_dirnames']]\n",
    "\n",
    "\n",
    "settings_aligned_umap_W12 = {\n",
    "    'gamma_list': gamma_list_v1,\n",
    "    'umap_kwargs': UMAP_KWARGS_DEFAULT,\n",
    "    'alignment_wrapper_kwargs': \n",
    "        {'subsample_ndim': None, \n",
    "         'subsample_ensemble': None,\n",
    "         'smod_last': True,\n",
    "         'use_01': False,\n",
    "         'skip_alignment': False},\n",
    "    'subplot_selection': [0, 3, 4, 5, 7, 8, 9, 13]    \n",
    "}\n",
    "settings_aligned_umap_W12['manyruns_dirnames'] = [\n",
    "    'W12_gamma%.2f_10k_periodic_R1_p3_M100_machineEps' % a \n",
    "    for a in gamma_list_v1]\n",
    "settings_aligned_umap_W12['manyruns_paths'] = [\n",
    "    RUNS_FOLDER + os.sep + 'multicell_manyruns' + os.sep + dirname \n",
    "    for dirname in settings_aligned_umap_W12['manyruns_dirnames']]\n",
    "\n",
    "\n",
    "settings_aligned_umap_W13 = {\n",
    "    'gamma_list': gamma_list_v1,\n",
    "    'umap_kwargs': UMAP_KWARGS_DEFAULT,\n",
    "    'alignment_wrapper_kwargs': \n",
    "        {'subsample_ndim': None, \n",
    "         'subsample_ensemble': None,\n",
    "         'smod_last': True,\n",
    "         'use_01': False,\n",
    "         'skip_alignment': False},\n",
    "    'subplot_selection': [0, 3, 4, 5, 7, 8, 9, 13]    \n",
    "}\n",
    "settings_aligned_umap_W13['manyruns_dirnames'] = [\n",
    "    'W13_gamma%.2f_10k_periodic_R1_p3_M100_machineEps' % a \n",
    "    for a in gamma_list_v1]\n",
    "settings_aligned_umap_W13['manyruns_paths'] = [\n",
    "    RUNS_FOLDER + os.sep + 'multicell_manyruns' + os.sep + dirname \n",
    "    for dirname in settings_aligned_umap_W13['manyruns_dirnames']]\n",
    "\n",
    "\n",
    "settings_aligned_umap_W14 = {\n",
    "    'gamma_list': gamma_list_v1,\n",
    "    'umap_kwargs': UMAP_KWARGS_DEFAULT,\n",
    "    'alignment_wrapper_kwargs': \n",
    "        {'subsample_ndim': None, \n",
    "         'subsample_ensemble': None,\n",
    "         'smod_last': True,\n",
    "         'use_01': False,\n",
    "         'skip_alignment': False},\n",
    "    'subplot_selection': [0, 3, 4, 5, 7, 8, 9, 13]    \n",
    "}\n",
    "settings_aligned_umap_W14['manyruns_dirnames'] = [\n",
    "    'W14_gamma%.2f_10k_periodic_R1_p3_M100_machineEps' % a \n",
    "    for a in gamma_list_v1]\n",
    "settings_aligned_umap_W14['manyruns_paths'] = [\n",
    "    RUNS_FOLDER + os.sep + 'multicell_manyruns' + os.sep + dirname \n",
    "    for dirname in settings_aligned_umap_W14['manyruns_dirnames']]\n",
    "\n",
    "\n",
    "settings_aligned_umap_W15 = {\n",
    "    'gamma_list': gamma_list_v1,\n",
    "    'umap_kwargs': UMAP_KWARGS_DEFAULT,\n",
    "    'alignment_wrapper_kwargs': \n",
    "        {'subsample_ndim': None, \n",
    "         'subsample_ensemble': None,\n",
    "         'smod_last': True,\n",
    "         'use_01': False,\n",
    "         'skip_alignment': False},\n",
    "    'subplot_selection': [0, 3, 4, 5, 7, 8, 9, 13]    \n",
    "}\n",
    "settings_aligned_umap_W15['manyruns_dirnames'] = [\n",
    "    'W15_gamma%.2f_10k_periodic_R1_p3_M100_machineEps' % a \n",
    "    for a in gamma_list_v1]\n",
    "settings_aligned_umap_W15['manyruns_paths'] = [\n",
    "    RUNS_FOLDER + os.sep + 'multicell_manyruns' + os.sep + dirname \n",
    "    for dirname in settings_aligned_umap_W15['manyruns_dirnames']]\n",
    "\n",
    "\"\"\"\n",
    "UMAP 8 panel configuration for each random W\n",
    "- orig (Wrandom0): seed=20 with rotation/flip:  'X_flip': True,  'X_rot': 310,\n",
    "- W1: seed 20\n",
    "- W2: seed 20\n",
    "- W7: seed 20\n",
    "- W12: seed 20\n",
    "- W13: seed 20\n",
    "- W14: seed 20\n",
    "- W15: seed 20\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# =========================================================\n",
    "settings_alignment_chosen = settings_aligned_umap_fig5_v3\n",
    "# =========================================================\n",
    "#settings_alignment_chosen = settings_aligned_umap_W15"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Choose settings option and embed/align UMAP for each manyruns**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "flag_generate = True  # If False, load data instead\n",
    "\n",
    "if flag_generate:\n",
    "    \n",
    "    # these asserts are needed for replotting the high dim states as tissues \n",
    "    #  using raw data within settings_alignment_chosen\n",
    "    assert not settings_alignment_chosen['alignment_wrapper_kwargs']['use_01']\n",
    "    assert settings_alignment_chosen['alignment_wrapper_kwargs']['subsample_ndim'] is None\n",
    "    assert settings_alignment_chosen['alignment_wrapper_kwargs']['subsample_ensemble'] is None\n",
    "\n",
    "    start = time.time()\n",
    "    embedded_datasets = gather_data_then_alignedUMAP(\n",
    "        settings_alignment_chosen['manyruns_paths'], \n",
    "        settings_alignment_chosen['manyruns_dirnames'],\n",
    "        settings_alignment_chosen['umap_kwargs'],\n",
    "        **settings_alignment_chosen['alignment_wrapper_kwargs']\n",
    "    )\n",
    "    end = time.time()\n",
    "    print('Total time (s):', end - start)\n",
    "    \n",
    "    # save embeddings and settings\n",
    "    # TODO - reimplement with new object embedded_datasets\n",
    "    #  - maybe best to save and load each individually? \n",
    "    \"\"\"pickled_alignUMAP = NOTEBOOK_OUTDIR + os.sep + 'aligned_mapper.pkl'\n",
    "    with open(pickled_alignUMAP, 'wb') as f:\n",
    "        # convert from class 'numba.typed.typedlist.List' to numpy array\n",
    "        raw_python_list = [i for i in aligned_mapper.embeddings_]\n",
    "        pickle.dump(raw_python_list, f)\n",
    "    \"\"\"\n",
    "    \n",
    "    # save settings \n",
    "    pickled_align_misc = NOTEBOOK_OUTDIR + os.sep + 'aligned_misc.pkl'\n",
    "    pickle.dump(settings_alignment_chosen, open(pickled_align_misc, 'wb'))\n",
    "    \n",
    "else:\n",
    "    # load settings\n",
    "    input_pickled_align_misc = NOTEBOOK_OUTDIR + os.sep + 'aligned_misc.pkl'\n",
    "    with open(input_pickled_alignUMAP, 'rb') as pickle_file:\n",
    "        settings_alignment_chosen = pickle.load(pickle_file)\n",
    "\n",
    "    # load main object (alignUMAP)\n",
    "    # TODO reimplement as above\n",
    "    \"\"\"\n",
    "    input_pickled_alignUMAP = NOTEBOOK_OUTDIR + os.sep + 'aligned_mapper.pkl'\n",
    "    with open(input_pickled_alignUMAP, 'rb') as pickle_file:\n",
    "        aligned_mapper_embeddings = pickle.load(pickle_file)\n",
    "    aligned_mapper = umap.AlignedUMAP(**settings_alignment_chosen['umap_kwargs'])\n",
    "    aligned_mapper.embeddings_ = aligned_mapper_embeddings    \n",
    "    \"\"\"\n",
    "    # embedded_datasets = ...LOAD each data_subdict object...\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "embedded_datasets = add_PCA_to_embedded_datasets(embedded_datasets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#for key in embedded_datasets.keys():\n",
    "#    embedded_datasets[key]['algos'].pop('pca')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part A: Generate 8 panel figure analagous to Fig. 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not NOTEBOOK_SPEEDRUN:\n",
    "    gamma_selection = settings_alignment_chosen['subplot_selection']  # [0, 3, 4, 5, 7, 8, 9, 13]\n",
    "    fpath = NOTEBOOK_OUTDIR + os.sep + 'picks_nobound_2x4_gammas%d' \\\n",
    "        % (len(settings_alignment_chosen['gamma_list']))\n",
    "\n",
    "    plot_args = (\n",
    "        embedded_datasets, \n",
    "        settings_alignment_chosen, \n",
    "    )\n",
    "    plot_kwargs = {\n",
    "        'annotate_indices': [],\n",
    "        'ext': ['.jpg', '.pdf', '.svg'],\n",
    "        'rasterized': True,\n",
    "        'ss': 4.0,  # 6\n",
    "        'edges': False,\n",
    "        'alpha': 1.0,\n",
    "        'black_edges': True,\n",
    "        'skip_cbar': False,\n",
    "        'X_flip': True,\n",
    "        'X_rot': 310,\n",
    "    }\n",
    "\n",
    "    c = 'energies'\n",
    "    fpath_spec = fpath + '_energies'\n",
    "    plot_aligned_panels_as_grid(*plot_args, fpath_spec, **plot_kwargs, c=c)\n",
    "\n",
    "    c = 'nunique'\n",
    "    fpath_spec = fpath + '_nunique'\n",
    "    plot_aligned_panels_as_grid(*plot_args, fpath_spec, **plot_kwargs, c=c)\n",
    "\n",
    "    #c = None\n",
    "    #fpath_spec = fpath + '_fixedcolor'\n",
    "    #plot_aligned_panels_as_grid(*plot_args, fpath_spec, **plot_kwargs, c=c)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part B: Plot distributions of fixed point data (energy, num. unique single cell states) on similar subplot grid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_multi_kwargs = settings_alignment_chosen['alignment_wrapper_kwargs'].copy()\n",
    "X_multi_kwargs.pop('skip_alignment', None)\n",
    "X_multi, num_cells, num_genes = build_X_multi(settings_alignment_chosen['manyruns_paths'], **X_multi_kwargs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fpath = NOTEBOOK_OUTDIR + os.sep + 'picks_nobound_2x4_gammas%d' \\\n",
    "    % (len(settings_alignment_chosen['gamma_list']))\n",
    "        \n",
    "plot_kwargs = {\n",
    "    'nrow': 2, \n",
    "    'ncol': 4, \n",
    "    'ext': ['.jpg', '.pdf'],\n",
    "    'gamma_selection': None\n",
    "}\n",
    "\n",
    "for fpdata_type in ['energies', 'nunique']:\n",
    "    fpath_figure = fpath + '_%sHist' % fpdata_type\n",
    "    plot_args = (\n",
    "        embedded_datasets,\n",
    "        settings_alignment_chosen, \n",
    "        fpdata_type, \n",
    "        fpath_figure\n",
    "    )\n",
    "    plot_fpdata_as_grid(*plot_args, **plot_kwargs, logy=False)\n",
    "    plot_fpdata_as_grid(*plot_args, **plot_kwargs, logy=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part C: Plotly style interactive plot of particular subpanel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from multicell.unsupervised_helper import \\\n",
    "    plotly_express_embedding, generate_control_data, plot_given_multicell, make_dimreduce_object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not NOTEBOOK_SPEEDRUN:\n",
    "\n",
    "    gamma_selection = settings_alignment_chosen['subplot_selection']\n",
    "    gamma_list = settings_alignment_chosen['gamma_list']\n",
    "\n",
    "    for idx in gamma_selection:\n",
    "        data_subdict_selected = embedded_datasets[idx]\n",
    "        outdir = NOTEBOOK_OUTDIR + os.sep + 'plotly_idx%d_gamma%.3f' % (idx, gamma_list[idx])\n",
    "        os.makedirs(outdir, exist_ok=True)\n",
    "\n",
    "        plotly_express_embedding(\n",
    "            data_subdict_selected, \n",
    "            color_by_index=False, \n",
    "            as_landscape=False, \n",
    "            fmod='jupyter', \n",
    "            show=False, \n",
    "            dirpath=outdir, \n",
    "            surf=False, \n",
    "            step=None\n",
    "        )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part D: Plot high-demensional state visualization of certain embedded points "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notes from powerpoint figure 5 (From bottom left):\n",
    "- BL: 1395, 3717, 2325, \n",
    "- TL: 2357, 728,\n",
    "- Homogenous + Maze: 636, 3473, \n",
    "- Heterogeneous: 3964, 1135, 1467, 2602"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# constants\n",
    "gamma_list = settings_alignment_chosen['gamma_list']\n",
    "assert gamma_list[13] == 1.0  # assumption for the manually created dicts below\n",
    "subset_orig = [1395, 3717, 2325, 2357, 728, 636, 3473, 3964, 1135, 1467, 2602]\n",
    "\n",
    "# dictionary of points to plot for each data_subdict (i.e. each gamma value UMAP)\n",
    "agg_points_to_plot_v1 = {\n",
    "    i: subset_orig for i in range(len(gamma_list))\n",
    "}\n",
    "\n",
    "agg_points_to_plot_v2 = {\n",
    "    0: [0, 1, 2, 3],\n",
    "    3: [8011, 5709, 8249, 6626, 9345, 9017, 8573, 9229],\n",
    "    4: [9521, 9904, 9691, 5542, 5402, 8322, 7098, 9852, 9844, 8186],\n",
    "    5: [9521, 9904, 9691, 5542, 5402, 8322, 7098, 9852, 9844, 8186],\n",
    "    7: [3886, 1551, 3649, 5492, 3809, 2990, 83, 1787, 8007, 7342, 7882, 4307],\n",
    "    8: [3886, 1551, 3649, 5492, 3809, 2990, 83, 1787, 8007, 7342, 7882, 4307, 9168, 775, 1512, 37],\n",
    "    9: [3886, 1551, 3649, 5492, 3809, 2990, 83, 1787, 8007, 7342, 7882, 4307, 9168, 775, 1512, 37],\n",
    "    13: subset_orig,  # Panel 8: original points\n",
    "}\n",
    "\n",
    "agg_points_to_plot_v3_many = {\n",
    "    13: [2269, 7142,\n",
    "         7350, 7694,\n",
    "         6809, 2, 2947,\n",
    "         8552,         \n",
    "         9713, 369, 52, 2970, 8237, 6789, \n",
    "         6681, 9018, 2722, 2503, 503, 3895, 7831, \n",
    "         9385, 1347, 2458, 5189, 7057,\n",
    "         9742, 5113,\n",
    "         8891, 1908, 2717, 3515, 5112, 5592, 7105, 9447, 9660,\n",
    "         1842, 8609, 4302, 6247, 5649, 224, 5776, 2504, 2978, 5727,\n",
    "         7657,\n",
    "         8134, 3657, 7849,\n",
    "         646, 5607, 3132, 4865, 7076,\n",
    "         5552, 8601, 1828, 2258, 3321,\n",
    "         6467, 2220, 3611, 7414, 5522\n",
    "        ]\n",
    "}\n",
    "\n",
    "agg_points_to_plot_v4_selected = {\n",
    "    13: [2269, 7694, 2,           # maze 'A'\n",
    "         6681, 5189, 5113,        # triple/homog 'B'\n",
    "         2717,                    # \"C\"\n",
    "         2978,                    # \"D\"\n",
    "         7657,                    # B'\n",
    "         6467, 5522               # A'\n",
    "        ]\n",
    "}\n",
    "\n",
    "agg_points_to_plot_v5_selected = {\n",
    "    13: [6006,              # -- image 1/11 (triple B, homog)\n",
    "         5883,              # -- image 2/11 (hetero \"C\", horizontal)\n",
    "         156,               # -- image 3/11 (maze A node)\n",
    "         5791,              # -- image 4/11 (maze A left)\n",
    "         8221,              # -- image 5/11 (maze A node)\n",
    "         2489,              # -- image 6/11 (triple B', ring 1)\n",
    "         8490,              # -- image 7/11 (triple B', mid lobe)\n",
    "         8278,              # -- image 8/11 (triple B', homog)\n",
    "         7354,              # -- image 9/11 (hetero \"D\", vertical)\n",
    "         4753,              # -- image 10/11 (maze A right)\n",
    "         598,               # -- image 11/11 (maze A node)\n",
    "         2058,              # -- extra image from disordered ring\n",
    "        ]\n",
    "}\n",
    "\n",
    "fmod = ''\n",
    "smod = settings_alignment_chosen['alignment_wrapper_kwargs']['smod_last']\n",
    "\n",
    "agg_points_to_plot = agg_points_to_plot_v5_selected\n",
    "\n",
    "# TODO implement this block -- diff points to plot for each gamma value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not NOTEBOOK_SPEEDRUN:\n",
    "\n",
    "    from multicell.multicell_replot import plot_tissue_given_agg_idx\n",
    "\n",
    "    for gamma_idx in agg_points_to_plot.keys():\n",
    "        gamma_val = gamma_list[gamma_idx]\n",
    "        outdir = NOTEBOOK_OUTDIR + os.sep + 'tissue_g%.3f' % gamma_val\n",
    "        os.makedirs(outdir, exist_ok=True)\n",
    "        data_subdict = embedded_datasets[gamma_idx]\n",
    "\n",
    "        print('Plotting points for subdict #%d' % gamma_idx)\n",
    "        agg_indices = agg_points_to_plot[gamma_idx]\n",
    "        for agg_idx in agg_indices:\n",
    "            title = 'Point: %d' % agg_idx  # None\n",
    "            plot_tissue_given_agg_idx(\n",
    "                data_subdict, agg_idx, fmod, outdir, \n",
    "                state_int=False, smod_last=smod, title=title)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part E: Plot 'annotated' version of alignedUMAP where some elements are circled"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subset_orig = [1395, 3717, 2325, 2357, 728, 636, 3473, 3964, 1135, 1467, 2602]\n",
    "fpath_annotate = fpath + '_annotate'\n",
    "ss = 6\n",
    "\n",
    "#agg_points_to_annotate = subset_orig\n",
    "agg_points_to_annotate = agg_points_to_plot_v3_many[13]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not NOTEBOOK_SPEEDRUN:\n",
    "\n",
    "    plot_args = (\n",
    "        embedded_datasets, \n",
    "        settings_alignment_chosen, \n",
    "        fpath_annotate\n",
    "    )\n",
    "    plot_kwargs = {\n",
    "        'ss': ss,\n",
    "        'annotate_indices': agg_points_to_annotate,\n",
    "        'addtext': True,\n",
    "        'X_flip': True,\n",
    "        'X_rot': 310,\n",
    "    }\n",
    "\n",
    "    plot_aligned_panels_as_grid(*plot_args, **plot_kwargs, ext=['.pdf', '.jpg'], rasterized=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part F: Plot single panel umap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gamma_idx = 13\n",
    "gamma_val = settings_alignment_chosen['gamma_list'][gamma_idx]\n",
    "assert gamma_val == 1.0\n",
    "fpath = NOTEBOOK_OUTDIR + os.sep + 'umap_singlepanel_g%.3f' % gamma_val\n",
    "\n",
    "annotate_indices = agg_points_to_plot_v5_selected[13] #[]\n",
    "color_choice = 'nunique' # None, 'energies', 'nunique'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not NOTEBOOK_SPEEDRUN:\n",
    "\n",
    "    plot_single_umap_panel_gridspec(\n",
    "        embedded_datasets[gamma_idx], fpath + '_notext', \n",
    "        ext=['.jpg', '.pdf', '.svg'], \n",
    "        ss=10, alpha=1.0, edges=False,\n",
    "        annotate_indices=annotate_indices, \n",
    "        rasterized=True,\n",
    "        c=color_choice,\n",
    "        skip_cbar=False,\n",
    "        X_flip=True, X_rot=310,\n",
    "        addtext=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "if not NOTEBOOK_SPEEDRUN:\n",
    "\n",
    "    plot_single_umap_panel_gridspec(\n",
    "        embedded_datasets[gamma_idx], fpath + '_energies', \n",
    "        ext=['.jpg', '.pdf', '.svg'], \n",
    "        ss=6, alpha=1.0, edges=False,\n",
    "        annotate_indices=[], \n",
    "        rasterized=True,\n",
    "        c='energies',\n",
    "        skip_cbar=False,\n",
    "        X_flip=True, X_rot=310,\n",
    "        addtext=False,\n",
    "        cbar_horizontal=False,\n",
    "        figsize=(5.6,3.2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not NOTEBOOK_SPEEDRUN:\n",
    "\n",
    "    plot_single_umap_panel_gridspec(\n",
    "        embedded_datasets[gamma_idx], fpath + '_addtext', \n",
    "        ext=['.jpg', '.pdf', '.svg'], \n",
    "        ss=10, alpha=1.0, edges=False,\n",
    "        annotate_indices=annotate_indices, \n",
    "        rasterized=True,\n",
    "        c=color_choice,\n",
    "        skip_cbar=False,\n",
    "        X_flip=True, X_rot=310,\n",
    "        addtext=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not NOTEBOOK_SPEEDRUN:\n",
    "\n",
    "    plot_single_umap_panel_gridspec(\n",
    "        embedded_datasets[gamma_idx], fpath + '_noColor_notext', \n",
    "        ext=['.jpg', '.pdf', '.svg'], \n",
    "        ss=10, alpha=1.0, edges=False,\n",
    "        annotate_indices=[], \n",
    "        rasterized=True,\n",
    "        c=None,\n",
    "        skip_cbar=False,\n",
    "        X_flip=True, X_rot=310,\n",
    "        addtext=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part G: Plot PCA grid and cumsum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gamma_selection = settings_alignment_chosen['subplot_selection']\n",
    "gamma_list = settings_alignment_chosen['gamma_list']\n",
    "fpath_pca_grid = NOTEBOOK_OUTDIR + os.sep + 'pca_picks_2x4'\n",
    "fpath_pca_cumsum = NOTEBOOK_OUTDIR + os.sep + 'pca_picks_cumsum'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*PCA grid in the style of aligned UMAP grid*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_args = (\n",
    "    embedded_datasets, \n",
    "    settings_alignment_chosen, \n",
    ")\n",
    "plot_kwargs = {\n",
    "    'algo': 'pca',\n",
    "    'annotate_indices': [],\n",
    "    'ext': ['.jpg', '.pdf', '.svg'],\n",
    "    'rasterized': True,\n",
    "    'ss': 6.0,\n",
    "    'edges': False,\n",
    "    'alpha': 1.0,\n",
    "    'black_edges': True,\n",
    "    'skip_cbar': False,\n",
    "}\n",
    "\n",
    "c = 'energies'\n",
    "fpath_spec = fpath_pca_grid + '_energies'\n",
    "plot_aligned_panels_as_grid(*plot_args, fpath_spec, **plot_kwargs, c=c)\n",
    "\n",
    "c = 'nunique'\n",
    "fpath_spec = fpath_pca_grid + '_nunique'\n",
    "plot_aligned_panels_as_grid(*plot_args, fpath_spec, **plot_kwargs, c=c)\n",
    "\n",
    "#c = None\n",
    "#fpath_spec = fpath + '_fixedcolor'\n",
    "#plot_aligned_panels_as_grid(*plot_args, fpath_spec, **plot_kwargs, c=c)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*PCA cumsum*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# full: (5, 3.5)\n",
    "# inset with xlim(0,25): (2.5, 2.5)\n",
    "fig = plt.figure(figsize=(5, 3.5), dpi=450)\n",
    "for idx in gamma_selection:\n",
    "    gamma_val = gamma_list[idx]\n",
    "    data_subdict_selected = embedded_datasets[idx]\n",
    "    pca_dict = data_subdict_selected['algos']['pca']\n",
    "    plt.plot(\n",
    "        pca_dict['cumsum_x'], \n",
    "        pca_dict['cumsum_y'], \n",
    "        '--',\n",
    "        label=r'$\\gamma=%.2f$' % gamma_val)\n",
    "\n",
    "plt.axvline(300, linestyle='-.', c='k', zorder=1)    \n",
    "\n",
    "# axis modifications\n",
    "#plt.gca().set_xscale('log')\n",
    "#plt.gca().set_yscale('log')\n",
    "#plt.xlim(0,25)\n",
    "\n",
    "#plt.legend()\n",
    "plt.xlabel('Num. components')\n",
    "plt.ylabel('Cumulative sum of eigenvalues')\n",
    "\n",
    "#plt.gcf().subplots_adjust(bottom=0.25, left=0.25)\n",
    "plt.gcf().subplots_adjust(bottom=0.15, left=0.25)\n",
    "plt.savefig(fpath_pca_cumsum + '.svg')\n",
    "plt.savefig(fpath_pca_cumsum + '.pdf')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part H: Correlation between various features"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Correlation: self-energy and interaction energy*\n",
    "- use violin plot? \n",
    "- energies array has 5 columns:\n",
    "  - 0: H_quadratic_form (should match index 1)\n",
    "  - 1: H_multi (which is sum of index (2,3,4)\n",
    "  - 2: H_self\n",
    "  - 3: H_app\n",
    "  - 4: H_pairwise_scaled\n",
    "- interesting: compare term 2 and 4 as $\\gamma$ changes\n",
    "  - can use dual violin plot https://seaborn.pydata.org/generated/seaborn.violinplot.html"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Correlation: energy and num unique single cell states*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pandas import DataFrame as df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fpath = NOTEBOOK_OUTDIR + os.sep + 'picks_nobound_2x4_gammas%d' \\\n",
    "    % (len(settings_alignment_chosen['gamma_list']))\n",
    "        \n",
    "plot_kwargs = {\n",
    "    'nrow': 2, \n",
    "    'ncol': 4, \n",
    "    'ext': ['.jpg', '.pdf'],\n",
    "    'gamma_selection': None\n",
    "}\n",
    "\n",
    "\n",
    "fpdata_type = ['nunique', 'energies']\n",
    "fpath_figure = fpath + '_nunique_vs_energies'\n",
    "plot_args = (\n",
    "    embedded_datasets,\n",
    "    settings_alignment_chosen, \n",
    "    fpdata_type, \n",
    "    fpath_figure\n",
    ")\n",
    "plot_fpdata_as_grid(*plot_args, **plot_kwargs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gamma_idx = 7  # 13 is 1.0\n",
    "gamma_value = settings_alignment_chosen['gamma_list'][gamma_idx]\n",
    "print('Selected gamma idx %d (value = %.3f)' % (gamma_idx, gamma_value))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "xx = embedded_datasets[gamma_idx]['nunique']\n",
    "yy = embedded_datasets[gamma_idx]['energies'][:,0]\n",
    "plt.figure(figsize=(4,4))\n",
    "#plt.title(r'Energy vs nunique for $\\gamma=%.2f$ (idx %d)' % \n",
    "#          (settings_alignment_chosen['gamma_list'][gamma_idx], gamma_idx))\n",
    "plt.xlabel('Num. unique s.c. states')\n",
    "plt.ylabel('Energy')\n",
    "plt.scatter(xx,yy,s=10,alpha=0.5)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Joint plot of energy and nunique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib.ticker import (MultipleLocator, AutoMinorLocator, LogLocator, NullFormatter)\n",
    "\n",
    "sns.set(style='white', \n",
    "        context='notebook', \n",
    "        rc={'figure.figsize':(14,10),\n",
    "            'xtick.bottom' : True, \n",
    "            'ytick.left' : True})\n",
    "\n",
    "def jointplot_energy_nunique(\n",
    "    gamma_idx, \n",
    "    add_spectrum_rescaled=False, \n",
    "    small=False,\n",
    "    logscale=True,\n",
    "    fmod=''):\n",
    "\n",
    "    gamma_value = settings_alignment_chosen['gamma_list'][gamma_idx]\n",
    "\n",
    "    # create dataframe for seaborn\n",
    "    xx = embedded_datasets[gamma_idx]['nunique']\n",
    "    yy = embedded_datasets[gamma_idx]['energies'][:,0]\n",
    "    custom_df = df(\n",
    "        {'Energy': yy, \n",
    "         'Num. unique single cell states': xx\n",
    "        }\n",
    "    )  \n",
    "    \n",
    "    \n",
    "    height = 5\n",
    "    size = 8\n",
    "    lw_jointgrid = 0.5\n",
    "    if small:\n",
    "        height = 3\n",
    "        size = 5\n",
    "        lw_jointgrid = 0.4\n",
    "    \n",
    "    g = sns.JointGrid(\n",
    "        x=\"Num. unique single cell states\", y=\"Energy\", \n",
    "        data=custom_df,\n",
    "        height=height,\n",
    "        ratio=2,\n",
    "        space=0.1,\n",
    "        marginal_ticks=True)\n",
    "\n",
    "    #marginal_kws=dict(stat='proportion'))\n",
    "    g = g.plot_joint(plt.scatter,  # plt.scatter, sns.kdeplot\n",
    "                     color=\"b\", \n",
    "                     edgecolor=\"white\", \n",
    "                     linewidth=0.5,\n",
    "                     marker='o', \n",
    "                     alpha=0.4,\n",
    "                     rasterized=True,\n",
    "                     s=size)\n",
    "\n",
    "    #g = g.plot_marginals(sns.kdeplot, shade=True)\n",
    "    nmax = np.max(xx)\n",
    "    nbin_max = max(38.5, nmax + 1.5)\n",
    "    \n",
    "    # add x axis labels manually\n",
    "    xticks = [0, 10, 20, 30]\n",
    "    if nmax >= 40:\n",
    "        xticks = np.arange(0,nmax,10)\n",
    "    g.ax_joint.set_xticks(xticks)\n",
    "    g.ax_joint.set_xticklabels(['%d' % a for a in xticks])\n",
    "\n",
    "    _ = g.ax_marg_x.hist(custom_df[\"Num. unique single cell states\"], \n",
    "                         color=\"r\", alpha=1, fill=False, ec='b',\n",
    "                         orientation=\"vertical\",\n",
    "                         bins=np.arange(0.5, nbin_max, 1))\n",
    "\n",
    "    _ = g.ax_marg_y.hist(custom_df[\"Energy\"], \n",
    "                         color=\"r\", alpha=1, fill=False, ec='b',\n",
    "                         orientation=\"horizontal\",\n",
    "                         bins=40)\n",
    "\n",
    "    if add_spectrum_rescaled:\n",
    "        multicell = embedded_datasets[gamma_idx]['multicell_template']\n",
    "        states = embedded_datasets[gamma_idx]['data']\n",
    "        J_multicell = multicell.matrix_J_multicell\n",
    "        ndim = J_multicell.shape[0]\n",
    "        rescaling_constant = - 2.0 / multicell.num_genes\n",
    "\n",
    "        # get spectrum of J_multicell\n",
    "        force_symm = False\n",
    "        if force_symm:\n",
    "            J_multicell = 0.5 * (J_multicell.T + J_multicell)\n",
    "        evals, evecs = sorted_eig(J_multicell, take_real=True, big_to_small=True)\n",
    "        eval_indices =  np.arange(len(evals))\n",
    "        print('states.shape', states.shape)\n",
    "        print('evecs.shape', evecs.shape)\n",
    "\n",
    "        # rescale spectrum\n",
    "        #npts = embedded_datasets[gamma_idx]['energies'].shape[0]\n",
    "        #energy_rescaled = embedded_datasets[gamma_idx]['energies'][:, 0] * rescaling_constant\n",
    "        J_multicell_spectrum_rescaled = evals / rescaling_constant\n",
    "\n",
    "        _ = g.ax_marg_y.hist(J_multicell_spectrum_rescaled, \n",
    "                             color=\"r\", alpha=1, fill=False, ec='r',\n",
    "                             orientation=\"horizontal\",\n",
    "                             bins=40)\n",
    "\n",
    "    assert len(custom_df.index) == 10000\n",
    "    \n",
    "    if logscale:\n",
    "        fmod += '_log'\n",
    "        g.ax_marg_x.set_yscale('log')\n",
    "        g.ax_marg_y.set_xscale('log')\n",
    "    \n",
    "        tick_pts_upper = [1, 10, 100, 1000, 10000]\n",
    "        tick_labels_upper = [r'$10^{-4}$', r'$10^{-3}$', r'$10^{-2}$', r'$10^{-1}$', r'$10^{0}$']\n",
    "        g.ax_marg_x.set_yticks(tick_pts_upper)\n",
    "        g.ax_marg_x.set_yticklabels(tick_labels_upper)\n",
    "\n",
    "        tick_pts_right = [1, 10, 100, 1000, 10000]\n",
    "        tick_labels_right = [r'$10^{-4}$', r'$10^{-3}$', r'$10^{-2}$', r'$10^{-1}$', r'$10^{0}$']\n",
    "        g.ax_marg_y.set_xticks(tick_pts_right)\n",
    "        g.ax_marg_y.set_xticklabels(tick_labels_right)\n",
    "\n",
    "        # set axis limits\n",
    "        g.ax_marg_y.set_xlim(0.7,10000*1.05)\n",
    "        g.ax_marg_x.set_ylim(0.7,10000*1.05)\n",
    "\n",
    "        add_minor_logticks = True\n",
    "        if add_minor_logticks:\n",
    "            y_minor = LogLocator(\n",
    "                base = 10.0, \n",
    "                subs=(0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9), \n",
    "                numticks=100\n",
    "            )\n",
    "            g.ax_marg_y.xaxis.set_minor_locator(y_minor)\n",
    "            g.ax_marg_y.xaxis.set_minor_formatter(NullFormatter())\n",
    "\n",
    "            g.ax_marg_x.yaxis.set_minor_locator(y_minor)\n",
    "            g.ax_marg_x.yaxis.set_minor_formatter(NullFormatter())\n",
    "            g.ax_marg_x.tick_params(axis='x', which='minor')\n",
    "\n",
    "            g.ax_marg_x.tick_params(axis='y', which='minor', direction='in', width=0.35)\n",
    "            g.ax_marg_y.tick_params(axis='x', which='minor', direction='in', width=0.35)\n",
    "\n",
    "    g.ax_joint.grid(linewidth=lw_jointgrid)\n",
    "    g.ax_marg_y.grid(axis='x', linestyle='--', linewidth=0.5)\n",
    "    g.ax_marg_x.grid(axis='y', linestyle='--', linewidth=0.5)\n",
    "    \n",
    "    if small:\n",
    "        fmod += '_small'\n",
    "    \n",
    "    plt.savefig(\n",
    "        NOTEBOOK_OUTDIR + os.sep + 'jointplot_custom_gamma%.3f%s.svg' % (gamma_value, fmod), \n",
    "        bbox_inches='tight', dpi=GLOBAL_DPI * 2)\n",
    "    plt.savefig(\n",
    "        NOTEBOOK_OUTDIR + os.sep + 'jointplot_custom_gamma%.3f%s.pdf' % (gamma_value, fmod), \n",
    "        bbox_inches='tight', dpi=GLOBAL_DPI * 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for idx, gamma in enumerate(settings_alignment_chosen['gamma_list']):\n",
    "    print('Joint plot (energy, nunique) for gamma', idx, gamma, '...')\n",
    "    jointplot_energy_nunique(idx)\n",
    "    jointplot_energy_nunique(idx, small=True)\n",
    "    #jointplot_energy_nunique(idx, logscale=False)\n",
    "    #jointplot_energy_nunique(idx, small=True, logscale=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Correlation: init cond to final state*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.clf()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Correlation: energy and spectrum*\n",
    "- approach 1: rank plot of spectrum; overlay energy values (rescaled) as thin axhline with some alpha\n",
    "- approach 2: x axis is eigenvalue/rescaled energy, y is frequency -- plot histograms with small bins\n",
    "- approach 3: check how the observed minima project onto the eigenvectors (part a) \n",
    "- approach 3: check how the observed minima project onto the eigenvectors (part b) \n",
    "- approach 4: similar to 3, but report the numerical error of the linalg.eig call by showing the magnitude of the projection of all states onto the eigenbasis. the projection coefficients (squared) should sum to ndim=900, but they don't exactly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from singlecell.singlecell_linalg import sorted_eig\n",
    "\n",
    "def plot_evals_rank(ax, evals, gamma, eval_indices, zorder=10, x_max=None):\n",
    "    x_axis = eval_indices  # possible use instead np.arange(len(eval_indices)), and change the axis label\n",
    "    evals = evals[eval_indices]\n",
    "    if x_max is not None:\n",
    "        # given scalar x_stretch, strecth x_axis s.t. x_axis[-1] = x_stretch\n",
    "        assert eval_indices[0] == 0\n",
    "        scale_factor = x_max / x_axis[-1]\n",
    "        x_axis = x_axis * scale_factor\n",
    "    ax.plot(x_axis, evals, '--o', c='k', alpha=0.6, label='$\\gamma=%.2f$' % gamma, zorder=zorder)\n",
    "    ax.axhline(0)\n",
    "    \n",
    "def plot_evals_hist(ax, evals, gamma, nbins=120, zorder=10):\n",
    "    #plt.scatter(range(len(evals)), evals, marker='o', alpha=0.6, edgecolors='k', label='$\\gamma=%.2f$' % gamma)\n",
    "    ax.hist(evals, alpha=0.6, bins=nbins, label='$\\gamma=%.2f$' % gamma, zorder=zorder) \n",
    "\n",
    "def get_and_plot_evals(J_multicell, gamma, zorder=10):\n",
    "    evals, _ = sorted_eig(J_multicell, take_real=True, big_to_small=True)\n",
    "    #plt.scatter(range(len(evals)), evals, marker='o', alpha=0.6, edgecolors='k', label='$\\gamma=%.2f$' % gamma)\n",
    "    plt.plot(range(len(evals)), evals, '--o', c='k', alpha=0.6, label='$\\gamma=%.2f$' % gamma, zorder=zorder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#THRESHOLD_A = 0.025  # discard components smaller than this\n",
    "THRESHOLD_B = 0.975    # def 0.99; alternative: add components until we get to this value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compare_energy_and_spectra(gamma_idx, approach=1):\n",
    "    assert approach in [1, 2, 3, 4, 5]\n",
    "    \n",
    "    gamma_val = settings_alignment_chosen['gamma_list'][gamma_idx]\n",
    "    multicell = embedded_datasets[gamma_idx]['multicell_template']\n",
    "    states = embedded_datasets[gamma_idx]['data']\n",
    "    J_multicell = multicell.matrix_J_multicell\n",
    "    ndim = J_multicell.shape[0]\n",
    "    # rescale by -2/N = -(2/NM)*M since we scale energy in multicell_metrics.py by 1/M already\n",
    "    rescaling_constant = - 2.0 / (ndim) * multicell.matrix_A.shape[0]  \n",
    "\n",
    "    # get spectrum of J_multicell\n",
    "    force_symm = False\n",
    "    if force_symm:\n",
    "        J_multicell = 0.5 * (J_multicell.T + J_multicell)\n",
    "    evals, evecs = sorted_eig(J_multicell, take_real=True, big_to_small=True)\n",
    "    eval_indices =  np.arange(len(evals))\n",
    "    print('states.shape', states.shape)\n",
    "    print('evecs.shape', evecs.shape)\n",
    "    \n",
    "    # rescale energies\n",
    "    npts = embedded_datasets[gamma_idx]['energies'].shape[0]\n",
    "    energy_rescaled = embedded_datasets[gamma_idx]['energies'][:, 0] * rescaling_constant\n",
    "    \n",
    "    plt.figure(figsize=(6.5, 6.5))\n",
    "    ax = plt.gca()\n",
    "    misc_output = None\n",
    "    \n",
    "    if approach == 1:\n",
    "        xlim_max = 900\n",
    "        plot_evals_rank(ax, evals, gamma_val, eval_indices, x_max=xlim_max)\n",
    "        plt.axhline(0, color='k', linestyle='--')\n",
    "        \n",
    "        \"\"\"\n",
    "        # plot rescaled emergy\n",
    "        for k in range(npts):\n",
    "            plt.axhline(energy_rescaled[k], alpha=0.4, linewidth=0.1, zorder=1)\n",
    "\n",
    "        ax.set_xlabel(r'Rank of $\\lambda$')\n",
    "        ax.set_ylabel(r'$\\lambda$')\n",
    "        plt.title(r'Rescaled energy $-2/{NM} \\cdot E(x)$ vs. spectrum for $\\gamma=%.2f$ (idx %d)' % \n",
    "                  (settings_alignment_chosen['gamma_list'][gamma_idx], gamma_idx))\n",
    "        plt.xlim(0,60)\n",
    "        #plt.xlim(0,25)\n",
    "        plt.ylim(np.min(energy_rescaled) * 0.95, np.max(evals) * 1.05)\n",
    "        #plt.gcf().set_figheight(5.5); plt.gcf().set_figwidth(5.5)\n",
    "        \"\"\"\n",
    "        # plot rescaled emergy histogram        \n",
    "        plt.hist(energy_rescaled, \n",
    "                 alpha=0.75, bins=40, label='$-2/{NM} \\cdot E(x)$, $\\gamma=%.2f$' % gamma_val, \n",
    "                 color='blue', zorder=5, lw=0.5, density=False,\n",
    "                 log=False, \n",
    "                 orientation='horizontal')\n",
    "        ax.set_xlabel(r'Rank of $\\lambda$ (and rescaled energy $-2/{NM} \\cdot E(x)$ bin height)')\n",
    "        ax.set_ylabel(r'$\\lambda$')\n",
    "        \n",
    "        #ax.set_yscale('log')\n",
    "        #plt.xlim(0, 900)\n",
    "        \n",
    "        plt.xlim(0, 50)\n",
    "        plt.ylim(1, 2.5)\n",
    "        ax.legend()\n",
    "        \n",
    "    elif approach == 2:\n",
    "        plt.hist(evals, \n",
    "                 alpha=0.75, bins=100, label='Spectrum, $\\gamma=%.2f$' % gamma_val, \n",
    "                 color='k', zorder=10, lw=0.5, density=False)\n",
    "        plt.hist(energy_rescaled, \n",
    "                 alpha=0.75, bins=40, label='$-2/{NM} \\cdot E(x)$, $\\gamma=%.2f$' % gamma_val, \n",
    "                 color='blue', zorder=5, lw=0.5, density=False)\n",
    "        ax.set_xlabel(r'$\\lambda$')\n",
    "        ax.set_xlabel(r'$\\lambda$')\n",
    "        #ax.set_xlim(-100,100)\n",
    "\n",
    "        #ax.set_ylabel('Observations')\n",
    "        ax.set_yscale('log')\n",
    "        ax.legend()\n",
    "    elif approach == 3:\n",
    "        for k in range(5):\n",
    "            proj = np.dot(states[k, :], evecs)  # care transpose, eigenvectors are the columns of the array\n",
    "            proj_normed = proj ** 2 / ndim\n",
    "            \n",
    "            eval_indices_descending = np.argsort(proj_normed)[::-1]\n",
    "            proj_normed_sorted = proj_normed[eval_indices_descending]\n",
    "            xaxis = [str(a) for a in eval_indices_descending]\n",
    "            \n",
    "            plt.bar(xaxis, proj_normed_sorted, alpha=0.6)\n",
    "            print('Check magnitude of projection coefficients for state %d:' % k, np.dot(proj, proj))\n",
    "            plt.xlim(0,40)\n",
    "            plt.xlabel(r'Rank $i$ of $\\lambda_i$')\n",
    "            plt.ylabel(r'Projection coefficient $\\frac{1}{NM} c_i^2$')\n",
    "            plt.title('State index: %d' %k)\n",
    "            plt.show()\n",
    "\n",
    "    elif approach == 4:\n",
    "        # how to decide on threshold? when the sgn(sum c_i v_i) ======= the state with high probability\n",
    "\n",
    "        # Step 1: compute num_components and validation array (both len npts)\n",
    "        num_components = np.zeros(npts, dtype=int)\n",
    "        state_reconstructed = np.zeros(npts, dtype=int)\n",
    "        \n",
    "        all_proj = np.dot(states, evecs)  # care transpose, eigenvectors are the columns of the array\n",
    "        all_proj_normed = all_proj ** 2 / ndim\n",
    "        \n",
    "        npts_4 = 10000\n",
    "        assert npts_4 == npts\n",
    "        \n",
    "        for k in range(npts_4):\n",
    "            #proj = np.dot(states[k, :], evecs)  # care transpose, eigenvectors are the columns of the array\n",
    "            #proj_normed = proj ** 2 / ndim\n",
    "            proj = all_proj[k, :]\n",
    "            proj_normed = all_proj_normed[k, :]\n",
    "            \n",
    "            eval_indices_descending = np.argsort(proj_normed)[::-1]\n",
    "            proj_normed_sorted = proj_normed[eval_indices_descending]\n",
    "            \n",
    "            # Using 'threshold type B':\n",
    "            proj_normed_sorted_cumsum = np.cumsum(proj_normed_sorted)\n",
    "            threshold_index = np.searchsorted(proj_normed_sorted_cumsum,\n",
    "                                              THRESHOLD_B)\n",
    "            num_components[k] = threshold_index + 1\n",
    "            #print(np.sum(proj_normed_sorted[0:threshold_index]))\n",
    "            #print('vs', np.sum(proj_normed_sorted[0:threshold_index+1]))\n",
    "            #print(k, num_components[k])\n",
    "            \n",
    "            # Check if this is able to 'reconstruct' the state\n",
    "            eval_indices_desc_trunc = eval_indices_descending[0:num_components[k]]\n",
    "            \n",
    "            reconstruction_thresholded = np.sum(proj[eval_indices_desc_trunc] * evecs[:, eval_indices_desc_trunc], axis=1)\n",
    "            error_arr = states[k, :] - np.sign(reconstruction_thresholded)\n",
    "            if np.all(error_arr == 0):\n",
    "                state_reconstructed[k] = 1\n",
    "            else:\n",
    "                print('FAIL', set(error_arr), np.count_nonzero(error_arr))\n",
    "                \n",
    "                \n",
    "            # BUGCHECK DETAILED DATA PLOTS\n",
    "            \"\"\"\n",
    "            if k == 0:\n",
    "                print('For k = 0 specifically...')\n",
    "                print('num_components[k]', num_components[k])\n",
    "                plt.figure(figsize=(10,3))\n",
    "                im = plt.imshow(evecs[:, eval_indices_desc_trunc])\n",
    "                print(np.min(evecs[:, eval_indices_desc_trunc]), np.max(evecs[:, eval_indices_desc_trunc]))\n",
    "                plt.colorbar()\n",
    "                plt.show()\"\"\"\n",
    "        \n",
    "        #plt.bar(np.arange(npts), num_components, alpha=0.75, edgecolor='k')\n",
    "        #plt.xlabel('state index')\n",
    "        #plt.ylabel('Min. num. components $c_i$ st. $T < \\frac{1}{NM} \\sum_i c_i^2$ ($T=$%.3f)' % THRESHOLD_B)\n",
    "\n",
    "        plt.hist(num_components, bins=np.arange(0.5, np.max(num_components) + 0.5), alpha=0.75, edgecolor='k')\n",
    "        plt.title('Reconstruction of state vector from top $K$ eigenvector components $(\\gamma=%.2f)$' % gamma_val)\n",
    "        plt.ylabel('Observations')\n",
    "        plt.xlabel('Min. num. of projection components $c_i$ s.t. $T < {NM}^{-1} \\sum_i^K c_i^2$ $(T=%.3f)$' % THRESHOLD_B)\n",
    "        plt.gca().set_yscale('log')\n",
    "        \n",
    "        print('%d of %d states reconstructed using THRESHOLD_B = %.3f' % \n",
    "              (np.sum(state_reconstructed), npts_4, THRESHOLD_B)\n",
    "             )\n",
    "        misc_output = (num_components, state_reconstructed)\n",
    "        \n",
    "    else:\n",
    "        assert approach == 5\n",
    "        \"\"\"\n",
    "        #proj_all = np.dot(states, evecs)\n",
    "        proj_magnitude = np.linalg.norm(proj_all, axis=1)**2\n",
    "        errors = ndim - proj_magnitude\n",
    "        plt.bar(np.arange(npts), errors, alpha=0.75, edgecolor='k')\n",
    "        plt.axhline(0)\"\"\"      \n",
    "        \n",
    "        norms = np.linalg.norm(evecs, axis=0)**2\n",
    "        plt.bar(np.arange(ndim), 1 - norms, alpha=0.75, edgecolor='k')\n",
    "        plt.axhline(0)\n",
    "        plt.xlabel(r'Evec index')\n",
    "        plt.ylabel(r'$1 - ||v_i||^2$')\n",
    "\n",
    "    #plt.gcf().subplots_adjust(left=0.1)\n",
    "    plt.savefig(NOTEBOOK_OUTDIR + os.sep + 'compare_spectrum_energy_gamma%.2f_%d_approach%d.pdf' % \n",
    "                (gamma_val, gamma_idx, approach),\n",
    "               bbox_inches='tight')\n",
    "    return ax, misc_output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gamma_idx = 7  # idx 13 is gamma 1.0; idx 9 is gamma 0.4\n",
    "gamma_val = settings_alignment_chosen['gamma_list'][gamma_idx]\n",
    "print(gamma_val)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ax = compare_energy_and_spectra(gamma_idx, approach=1)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "ax = compare_energy_and_spectra(gamma_idx, approach=2)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ax = compare_energy_and_spectra(gamma_idx, approach=3)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "ax, misc_output = compare_energy_and_spectra(gamma_idx, approach=4)\n",
    "#plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "num_components = misc_output[0]\n",
    "state_reconstructed = misc_output[1]\n",
    "\n",
    "plt.figure(figsize=(6.5, 6.5))\n",
    "plt.hist(num_components, bins=np.arange(0.5, np.max(num_components) + 0.5), alpha=0.75, edgecolor='k')\n",
    "plt.title('Reconstruction of state vector from top $K$ eigenvector components $(\\gamma=%.2f)$' % gamma_val)\n",
    "plt.ylabel('Observations')\n",
    "plt.xlabel('Min. num. of projection components $c_i$ s.t. $T < {NM}^{-1} \\sum_i^K c_i^2$ $(T=%.3f)$' % THRESHOLD_B)\n",
    "\n",
    "#plt.gca().set_yscale('log')\n",
    "plt.xlim(0,40)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ax, misc_output = compare_energy_and_spectra(gamma_idx, approach=5)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#arr = np.eye(3)\n",
    "arr = np.array([\n",
    "    [0,1,1],\n",
    "    [1,1,1],\n",
    "    [1,0,1]\n",
    "])\n",
    "\n",
    "\n",
    "E,V = np.linalg.eig(arr)\n",
    "print(E)\n",
    "print(V)\n",
    "\n",
    "print(np.dot(arr, V[:,0]) - E[0] * V[:,0])\n",
    "print(np.dot(arr, V[0,:]) - E[0] * V[0,:])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot particular state as eigenvector decomposition in lattice format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_evec_as_lattice(evec, num_cells, num_genes, sidelength, fpath, title=None, norm=True):\n",
    "    ndim = num_cells * num_genes\n",
    "    assert sidelength ** 2 == num_cells\n",
    "    assert len(evec.shape) == 1\n",
    "    assert evec.shape[0] == ndim\n",
    "    \n",
    "    if norm:\n",
    "        arr_caps = np.amax(np.abs(evec), axis=0)  # could simplify, its not 2D array\n",
    "        if arr_caps != np.zeros_like(arr_caps):\n",
    "            evec = evec / arr_caps  # absmax of each eigenvector is stretched to +1    \n",
    "        vmin = -1.0\n",
    "        vmax = 1.0\n",
    "    else:\n",
    "        vmin = None\n",
    "        vmax = None\n",
    "            \n",
    "    \n",
    "    # morph array into correct shape (first lattice site has genes 1 to N, next has N+1 to N, etc.)\n",
    "    sidelength_genes = sidelength * num_genes\n",
    "    arr = evec.reshape(sidelength, sidelength, num_genes, order='C')\n",
    "    arr_imshow = np.zeros((sidelength_genes, sidelength))\n",
    "    for idx in range(sidelength):\n",
    "        i = num_genes * idx\n",
    "        j = num_genes * (idx + 1)\n",
    "        arr_imshow[i:j, :] = arr[idx, :, :].T\n",
    "    \n",
    "    plt.figure(figsize=(6,5))\n",
    "    ax = plt.gca()\n",
    "    mappable = ax.imshow(arr_imshow, interpolation='none', cmap='coolwarm_r', aspect='auto', \n",
    "                         vmin=vmin, vmax=vmax)\n",
    "    \n",
    "    # add line dividers for each cell\n",
    "    vlw = 2.0\n",
    "    hlw = 2.0\n",
    "    linecol = 'k'\n",
    "    for k in range(1, sidelength):\n",
    "        ax.axvline(k - 0.5, 0, 1, color=linecol, lw=vlw)\n",
    "        ax.axhline(k * num_genes  - 0.5, 0, 1, color=linecol, lw=hlw)\n",
    "    \n",
    "    if title is not None:\n",
    "        plt.title(title)\n",
    "    \n",
    "    plt.savefig(fpath + '.png', bbox_inches='tight')\n",
    "    plt.savefig(fpath + '.pdf', bbox_inches='tight')\n",
    "    \n",
    "\n",
    "def get_spectral_approx_of_state(state_idx, gamma_idx, num_components, show_proj=True):\n",
    "    gamma_val = settings_alignment_chosen['gamma_list'][gamma_idx]\n",
    "    multicell = embedded_datasets[gamma_idx]['multicell_template']\n",
    "    states = embedded_datasets[gamma_idx]['data']\n",
    "    J_multicell = multicell.matrix_J_multicell\n",
    "    ndim = J_multicell.shape[0]\n",
    "    rescaling_constant = - 2.0 / multicell.num_genes\n",
    "\n",
    "    # get spectrum of J_multicell\n",
    "    force_symm = False\n",
    "    if force_symm:\n",
    "        J_multicell = 0.5 * (J_multicell.T + J_multicell)\n",
    "    evals, evecs = sorted_eig(J_multicell, take_real=True, big_to_small=True)\n",
    "    eval_indices =  np.arange(len(evals))\n",
    "    print('states.shape', states.shape)\n",
    "    print('evecs.shape', evecs.shape)\n",
    "    \n",
    "    # algebra below            \n",
    "    proj = np.dot(states[state_idx, :], evecs)  # care transpose, eigenvectors are the columns of the array\n",
    "    proj_normed = proj ** 2 / ndim\n",
    "    \n",
    "    eval_indices_descending = np.argsort(proj_normed)[::-1]\n",
    "    proj_normed_sorted = proj_normed[eval_indices_descending]    \n",
    "    eval_indices_desc_trunc = eval_indices_descending[0:num_components]\n",
    "\n",
    "    state_reconstruction = np.sum(proj[eval_indices_desc_trunc] * evecs[:, eval_indices_desc_trunc], axis=1)\n",
    "    error_state = states[state_idx, :] - np.sign(state_reconstruction)\n",
    "    \n",
    "    if np.all(error_state == 0):\n",
    "        print('Perfectly reconstructed with %d components' % num_components)\n",
    "    else:\n",
    "        print('FAIL', set(error_state), 'total gene errors:', np.count_nonzero(error_state))\n",
    "        \n",
    "    if show_proj:\n",
    "        xaxis = [str(a) for a in eval_indices_descending]\n",
    "        plt.figure(figsize=(10,5))\n",
    "        plt.bar(xaxis, proj_normed_sorted, alpha=0.6)\n",
    "        print('Check magnitude of projection coefficients for state %d:' % state_idx, \n",
    "              np.dot(proj, proj))\n",
    "        plt.xlim(-0.5,20)\n",
    "        plt.xlabel(r'Rank $i$ of $\\lambda_i$')\n",
    "        plt.ylabel(r'Projection coefficient $\\frac{1}{NM} c_i^2$')\n",
    "        plt.title('State index: %d' % state_idx)\n",
    "        plt.show()\n",
    "                \n",
    "    return state_reconstruction, error_state "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#state_idx = 4854  # 8181 for gamma idx 7; 4854 for gamma idx 5\n",
    "#gamma_idx = 5\n",
    "\n",
    "state_idx = 8181  # 8181 for gamma idx 7; 4854 for gamma idx 5\n",
    "gamma_idx = 7\n",
    "\n",
    "gamma_val = gamma_list[gamma_idx]\n",
    "\n",
    "num_cells = 100\n",
    "num_genes = 9\n",
    "sidelength = 10\n",
    "\n",
    "outdir = NOTEBOOK_OUTDIR + os.sep + 'tissue_g%.3f_agg%d_spectralproj' % (gamma_val, state_idx)\n",
    "os.makedirs(outdir, exist_ok=True)\n",
    "print('gamma_val:', gamma_val)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####  Plot true state in fancy format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_subdict = embedded_datasets[gamma_idx]   \n",
    "title = 'Point: %d' % state_idx  # None\n",
    "plot_tissue_given_agg_idx(\n",
    "    data_subdict, state_idx, fmod, outdir, \n",
    "    state_int=False, smod_last=smod, title=title)\n",
    "print('plotted tissue in %s' % outdir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Approximate the state using top-k eigenvector weighted sum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "state_true = data_subdict['data'][state_idx, :]\n",
    "\n",
    "num_components = 2  # how many evecs to include in reconstruction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "state_reconstruction, reconstruction_error = get_spectral_approx_of_state(\n",
    "    state_idx, gamma_idx, num_components)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "title = 'True state %d for $\\gamma=%0.3f$' % (state_idx, gamma_val)\n",
    "plot_evec_as_lattice(state_true, num_cells, num_genes, sidelength, \n",
    "                     outdir + os.sep + 'truestate',\n",
    "                     norm=True,\n",
    "                     title=title)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "title = 'Approx. state %d for $\\gamma=%0.3f$ (components: %d)' % (state_idx, gamma_val, num_components)\n",
    "plot_evec_as_lattice(state_reconstruction, num_cells, num_genes, sidelength, \n",
    "                     outdir + os.sep + 'ncomponents%d_stateapprox' % num_components,\n",
    "                     norm=True,\n",
    "                     title=title)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "title = 'Approx. state (signed) %d for $\\gamma=%0.3f$ (components: %d)' % (state_idx, gamma_val, num_components)\n",
    "plot_evec_as_lattice(np.sign(state_reconstruction), num_cells, num_genes, sidelength, \n",
    "                     outdir + os.sep + 'ncomponents%d_stateapproxSigned' % num_components,\n",
    "                     norm=True,\n",
    "                     title=title)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "title = 'Approximation error for state %d for $\\gamma=%0.3f$ (components: %d)' % (state_idx, gamma_val, num_components)\n",
    "plot_evec_as_lattice(reconstruction_error, num_cells, num_genes, sidelength, \n",
    "                     outdir + os.sep + 'ncomponents%d_stateapprox_error' % num_components,\n",
    "                     norm=True,\n",
    "                     title=title)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "pycharm": {
   "stem_cell": {
    "cell_type": "raw",
    "metadata": {
     "collapsed": false
    },
    "source": []
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
